diff --git a/src/urma/lib/uvs/core/include/uvs_api.h b/src/urma/lib/uvs/core/include/uvs_api.h
index c699516c4936a118da1a34f89f723539d958a6f3..cddf20efce68dc013b0abf76f55ad5227b71d665 100644
--- a/src/urma/lib/uvs/core/include/uvs_api.h
+++ b/src/urma/lib/uvs/core/include/uvs_api.h
@@ -75,6 +75,13 @@ int uvs_get_device_name_by_eid(uvs_eid_t *eid, char *buf, size_t len);
  */
 int uvs_set_topo_info(void *topo, uint32_t topo_num);
 
+/**
+ * UVS get topo info.
+ * @param[out] topo: topo map
+ * Return: 0 on success, other value on error
+ */
+int uvs_get_topo_info(void *topo);
+
 /**
  * Get primary and port eid from topo info.
  * @param[in] route: parameter that contains src_v_eid and dst_v_eid,
diff --git a/src/urma/lib/uvs/core/tpsa_api.c b/src/urma/lib/uvs/core/tpsa_api.c
index 1f53b6ea7c0b5a0dc15c778223d7e4258ee9d357..d28c3186ac932b5961638ce109b6b1ac9d995069 100644
--- a/src/urma/lib/uvs/core/tpsa_api.c
+++ b/src/urma/lib/uvs/core/tpsa_api.c
@@ -99,6 +99,26 @@ int uvs_set_topo_info_inner(void *topo, uint32_t topo_num)
     return ret;
 }
 
+int uvs_get_topo_info_inner(void *topo)
+{
+    int ret;
+
+    if (!topo) {
+        TPSA_LOG_ERR("topo is NULL.\n");
+        return -EINVAL;
+    }
+
+    ret = uvs_ubcore_ioctl_get_topo(topo);
+    if (ret != 0) {
+        TPSA_LOG_ERR("failed to get topo info in ubcore.\n");
+        return ret;
+    } else {
+        TPSA_LOG_INFO("success to get topo info in ubcore\n");
+    }
+
+    return ret;
+}
+
 int uvs_set_topo_info(void *topo, uint32_t topo_num)
 {
     int ret = 0;
@@ -108,6 +128,15 @@ int uvs_set_topo_info(void *topo, uint32_t topo_num)
     return ret;
 }
 
+int uvs_get_topo_info(void *topo)
+{
+    int ret = 0;
+    uvs_get_api_rdlock();
+    ret = uvs_get_topo_info_inner(topo);
+    put_uvs_lock();
+    return ret;
+}
+
 int uvs_get_route_list(const uvs_route_t *route, uvs_route_list_t *route_list)
 {
     int ret = 0;
diff --git a/src/urma/lib/uvs/core/tpsa_ioctl.h b/src/urma/lib/uvs/core/tpsa_ioctl.h
index 12885e8d465a24b5c3d81cb13901881ef594b767..d4034cfa5a94a3079d6d56eee8162193708ebb0e 100644
--- a/src/urma/lib/uvs/core/tpsa_ioctl.h
+++ b/src/urma/lib/uvs/core/tpsa_ioctl.h
@@ -44,6 +44,7 @@ typedef struct tpsa_cmd_hdr {
 typedef enum uvs_global_cmd {
     UVS_CMD_SET_TOPO = 1,
     UVS_CMD_GET_TOPO_EID = 2,
+    UVS_CMD_GET_TOPO = 3,
     UVS_CMD_GLOBAL_LAST
 } uvs_global_cmd_t;
 
@@ -54,6 +55,12 @@ typedef struct uvs_set_topo {
     } in;
 } uvs_set_topo_t;
 
+typedef struct uvs_get_topo {
+    struct {
+        void *topo_map;
+    } out;
+} uvs_get_topo_t;
+
 typedef struct uvs_cmd_get_route_list {
     uvs_route_t in;
     uvs_route_list_t out;
@@ -61,6 +68,7 @@ typedef struct uvs_cmd_get_route_list {
 
 int uvs_ioctl_in_global(tpsa_ioctl_ctx_t *ioctl_ctx, uvs_global_cmd_t cmd, void *arg, uint32_t arg_len);
 int uvs_ioctl_set_topo(tpsa_ioctl_ctx_t *ioctl_ctx, uvs_set_topo_t *arg);
+int uvs_ioctl_get_topo(tpsa_ioctl_ctx_t *ioctl_ctx, uvs_get_topo_t *arg);
 int uvs_ioctl_get_route_list(tpsa_ioctl_ctx_t *ioctl_ctx, uvs_cmd_get_route_list_t *arg);
 
 #ifdef __cplusplus
diff --git a/src/urma/lib/uvs/core/uvs_cmd_tlv.c b/src/urma/lib/uvs/core/uvs_cmd_tlv.c
index 3aad77aa52e0bb5c0ba049c9d9b49887bc0840c0..796fcc0660a1599f57a8f9ecd14ba6c99a10263d 100644
--- a/src/urma/lib/uvs/core/uvs_cmd_tlv.c
+++ b/src/urma/lib/uvs/core/uvs_cmd_tlv.c
@@ -54,6 +54,16 @@ int uvs_ioctl_set_topo(tpsa_ioctl_ctx_t *ioctl_ctx, uvs_set_topo_t *arg)
     return uvs_ioctl_in_global(ioctl_ctx, UVS_CMD_SET_TOPO, (void *)attrs, sizeof(attrs));
 }
 
+int uvs_ioctl_get_topo(tpsa_ioctl_ctx_t *ioctl_ctx, uvs_get_topo_t *arg)
+{
+    uvs_cmd_attr_t attrs[GET_TOPO_OUT_NUM] = {0};
+    uvs_cmd_attr_t *a = attrs;
+
+    ATTR(a++, GET_TOPO_OUT_TOPO_MAP, arg->out.topo_map);
+
+    return uvs_ioctl_in_global(ioctl_ctx, UVS_CMD_GET_TOPO, (void *)attrs, sizeof(attrs));
+}
+
 int uvs_ioctl_get_route_list(tpsa_ioctl_ctx_t *ioctl_ctx, uvs_cmd_get_route_list_t *arg)
 {
     uvs_cmd_attr_t attrs[GET_ROUTE_LIST_IN_NUM + GET_ROUTE_LIST_OUT_NUM - UVS_CMD_OUT_TYPE_INIT] = {0};
diff --git a/src/urma/lib/uvs/core/uvs_cmd_tlv.h b/src/urma/lib/uvs/core/uvs_cmd_tlv.h
index d308f19dd837e0f3b3ff65d3d2f3c14d98cae61f..5803fdc2c810f3e4b7c8d75840e80b2febd8ef6f 100644
--- a/src/urma/lib/uvs/core/uvs_cmd_tlv.h
+++ b/src/urma/lib/uvs/core/uvs_cmd_tlv.h
@@ -42,6 +42,12 @@ typedef enum uvs_cmd_set_topo_type {
     SET_TOPO_IN_NUM /* Only for calculating number of types */
 } uvs_cmd_set_topo_type_t;
 
+typedef enum uvs_cmd_get_topo_type {
+    /* In type */
+    GET_TOPO_OUT_TOPO_MAP,
+    GET_TOPO_OUT_NUM /* Only for calculating number of types */
+} uvs_cmd_get_topo_type_t;
+
 typedef enum uvs_cmd_get_route_list_type {
     GET_ROUTE_LIST_IN_ROUTE_PAIR,
     GET_ROUTE_LIST_IN_NUM,
diff --git a/src/urma/lib/uvs/core/uvs_ubagg_ioctl.c b/src/urma/lib/uvs/core/uvs_ubagg_ioctl.c
index d33df8d0714b0a0c0f5a4635a026981d9a8837eb..8851f834f51dfef22a3f8ee4ff288af7b2edd1df 100644
--- a/src/urma/lib/uvs/core/uvs_ubagg_ioctl.c
+++ b/src/urma/lib/uvs/core/uvs_ubagg_ioctl.c
@@ -139,6 +139,32 @@ int uvs_ubagg_ioctl_set_topo(void *topo_info, int topo_num)
     return 0;
 }
 
+int uvs_ubcore_ioctl_get_topo(void *topo_map)
+{
+    tpsa_ioctl_ctx_t ioctl_ctx = {0};
+    uvs_get_topo_t arg = {0};
+    int ret = 0;
+
+    int dev_fd = open(UVS_UBCORE_DEVICE_PATH, O_RDWR);
+    if (dev_fd == -1) {
+        TPSA_LOG_ERR("Failed to open dev_fd err: %s.\n", ub_strerror(errno));
+        return -1;
+    }
+
+    ioctl_ctx.ubcore_fd = dev_fd;
+    arg.out.topo_map = topo_map;
+
+    ret = uvs_ioctl_get_topo(&ioctl_ctx, &arg);
+    if (ret != 0) {
+        TPSA_LOG_ERR("uvs_ubcore_ioctl_get_topo fail\n");
+        close(dev_fd);
+        return -1;
+    }
+
+    close(dev_fd);
+    return 0;
+}
+
 int uvs_ubcore_ioctl_set_topo(void *topo_info, int topo_num)
 {
     tpsa_ioctl_ctx_t ioctl_ctx = {0};
diff --git a/src/urma/lib/uvs/core/uvs_ubagg_ioctl.h b/src/urma/lib/uvs/core/uvs_ubagg_ioctl.h
index 9f327279bd1881534d2b16fb80f65b8a6e339cf1..cdf23b2a35e46abd5437b4049dc0bc02e72cde32 100644
--- a/src/urma/lib/uvs/core/uvs_ubagg_ioctl.h
+++ b/src/urma/lib/uvs/core/uvs_ubagg_ioctl.h
@@ -64,6 +64,7 @@ int uvs_ubagg_ioctl_create_agg_dev(uvs_eid_t *agg_eid);
 int uvs_ubagg_ioctl_delete_agg_dev(uvs_eid_t *agg_eid);
 int uvs_ubagg_ioctl_get_dev_name_by_eid(uvs_eid_t *eid, char *buf, size_t len);
 int uvs_ubagg_ioctl_set_topo(void *topo_info, int topo_num);
+int uvs_ubcore_ioctl_get_topo(void *topo_map);
 int uvs_ubcore_ioctl_set_topo(void *topo_info, int topo_num);
 int uvs_ubcore_ioctl_get_route_list(const uvs_route_t *route, uvs_route_list_t *route_list);
 
diff --git a/src/urma/tools/CMakeLists.txt b/src/urma/tools/CMakeLists.txt
index 2f36ed570ab43dabafa5891125b830950085fba3..612731f89731c915af1654babc272ec18a108679 100644
--- a/src/urma/tools/CMakeLists.txt
+++ b/src/urma/tools/CMakeLists.txt
@@ -1,8 +1,7 @@
 # SPDX-License-Identifier: MIT
 # Copyright (c) Huawei Technologies Co., Ltd. 2021-2025. All rights reserved.
 
-if ((NOT BUILD_FOR_ANDROID) AND (NOT OHOS) AND (NOT BUILD_FOR_OH))
-    add_subdirectory(urma_admin)
-    add_subdirectory(urma_perftest)
-    add_subdirectory(ubagg_cli)
-endif ()
+add_subdirectory(urma_admin)
+add_subdirectory(urma_perftest)
+add_subdirectory(urma_ping)
+add_subdirectory(ubagg_cli)
diff --git a/src/urma/tools/urma_ping/CMakeLists.txt b/src/urma/tools/urma_ping/CMakeLists.txt
new file mode 100644
index 0000000000000000000000000000000000000000..4c34432a18e5975a3cc50ae528de897497253d4a
--- /dev/null
+++ b/src/urma/tools/urma_ping/CMakeLists.txt
@@ -0,0 +1,30 @@
+# SPDX-License-Identifier: MIT
+# Copyright (c) Huawei Technologies Co., Ltd. 2026-2026. All rights reserved.
+
+add_executable(urma_ping
+    ping_log.c
+    ping_parameters.c
+    ping_run.c
+    ping_stat.c
+    ping.c
+)
+
+set_target_properties(urma_ping
+    PROPERTIES
+        C_STANDARD 99
+)
+
+target_include_directories(urma_ping
+    PRIVATE
+        ${CMAKE_SOURCE_DIR}/urma/lib/uvs/core/include
+)
+
+target_link_libraries(urma_ping
+    PRIVATE
+        urma
+        tpsa
+)
+
+install(TARGETS urma_ping
+    DESTINATION /usr/bin
+)
diff --git a/src/urma/tools/urma_ping/ping.c b/src/urma/tools/urma_ping/ping.c
new file mode 100644
index 0000000000000000000000000000000000000000..d2345b16fa28769c32f460dd15882cf581178b94
--- /dev/null
+++ b/src/urma/tools/urma_ping/ping.c
@@ -0,0 +1,50 @@
+/*
+ * SPDX-License-Identifier: MIT
+ * Copyright (c) Huawei Technologies Co., Ltd. 2026-2026. All rights reserved.
+ * Description: urma_ping entry file
+ * Author: Wang Hang
+ * Create: 2026-02-02
+ * Note:
+ * History: 2026-02-02 Create file
+ */
+
+#include <errno.h>
+#include <libgen.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "ping_log.h"
+#include "ping_parameters.h"
+#include "ping_run.h"
+
+int main(int argc, char *argv[])
+{
+    ping_cfg_t cfg = {
+        .argc = argc,
+        .argv = argv,
+        .filename = basename(argv[0]),
+        .count = UINT32_MAX,
+        .interval = 1,
+        .size = 4,
+        .verbose_level = VLOG_LEVEL_NORMAL,
+        .deadline = 0,
+        .timeout = 1,
+    };
+
+    int ret = 0;
+
+    if ((ret = parse_args(&cfg)) != 0) {
+        return EXIT_FAILURE;
+    }
+
+    if ((ret = check_args(&cfg)) != 0) {
+        return EXIT_FAILURE;
+    }
+
+    if ((ret = start_ping(&cfg)) != 0) {
+        return EXIT_FAILURE;
+    }
+
+    return 0;
+}
diff --git a/src/urma/tools/urma_ping/ping_log.c b/src/urma/tools/urma_ping/ping_log.c
new file mode 100644
index 0000000000000000000000000000000000000000..cd677f1925452353470d79ec330239ecbf2307ec
--- /dev/null
+++ b/src/urma/tools/urma_ping/ping_log.c
@@ -0,0 +1,33 @@
+/*
+ * SPDX-License-Identifier: MIT
+ * Copyright (c) Huawei Technologies Co., Ltd. 2026-2026. All rights reserved.
+ * Description: urma_ping log implementation file
+ * Author: Wang Hang
+ * Create: 2026-02-03
+ * Note:
+ * History: 2026-02-03 Create file
+ */
+
+#include <stdarg.h>
+#include <stdio.h>
+
+#include "ping_log.h"
+
+static int verbose_level = VLOG_LEVEL_NORMAL;
+
+void verbose_set_level(ping_vlog_level_t level)
+{
+    verbose_level = level;
+}
+
+void verbose_print(FILE *stream, ping_vlog_level_t level, const char *fmt, ...)
+{
+    if (verbose_level < level) {
+        return;
+    }
+
+    va_list va;
+    va_start(va, fmt);
+    vfprintf(stream, fmt, va);
+    va_end(va);
+}
diff --git a/src/urma/tools/urma_ping/ping_log.h b/src/urma/tools/urma_ping/ping_log.h
new file mode 100644
index 0000000000000000000000000000000000000000..5bfda6618deab97e9983447e9f1d9005448133ee
--- /dev/null
+++ b/src/urma/tools/urma_ping/ping_log.h
@@ -0,0 +1,33 @@
+/*
+ * SPDX-License-Identifier: MIT
+ * Copyright (c) Huawei Technologies Co., Ltd. 2026-2026. All rights reserved.
+ * Description: urma_ping log head file
+ * Author: Wang Hang
+ * Create: 2026-02-03
+ * Note:
+ * History: 2026-02-03 Create file
+ */
+
+#ifndef URMA_PING_LOG_H
+#define URMA_PING_LOG_H
+
+#include <stdarg.h>
+#include <stdio.h>
+
+typedef enum ping_vlog_level {
+    VLOG_LEVEL_QUIET = 0,
+    VLOG_LEVEL_NORMAL = 1,
+    VLOG_LEVEL_VERBOSE = 2,
+    VLOG_LEVEL_VVERBOSE = 3,
+} ping_vlog_level_t;
+
+void verbose_set_level(ping_vlog_level_t level);
+void verbose_print(FILE *stream, ping_vlog_level_t level, const char *fmt, ...);
+
+#define LOG_QUIET(...)    verbose_print(stdout, VLOG_LEVEL_QUIET, __VA_ARGS__)
+#define LOG_NORMAL(...)   verbose_print(stdout, VLOG_LEVEL_NORMAL, __VA_ARGS__)
+#define LOG_VERBOSE(...)  verbose_print(stdout, VLOG_LEVEL_VERBOSE, __VA_ARGS__)
+#define LOG_VVERBOSE(...) verbose_print(stdout, VLOG_LEVEL_VVERBOSE, __VA_ARGS__)
+#define LOG_ERROR(...)    verbose_print(stderr, VLOG_LEVEL_QUIET, __VA_ARGS__)
+
+#endif
diff --git a/src/urma/tools/urma_ping/ping_parameters.c b/src/urma/tools/urma_ping/ping_parameters.c
new file mode 100644
index 0000000000000000000000000000000000000000..691a845a3f994ac5120f453faccac3d2812dd6fe
--- /dev/null
+++ b/src/urma/tools/urma_ping/ping_parameters.c
@@ -0,0 +1,188 @@
+/*
+ * SPDX-License-Identifier: MIT
+ * Copyright (c) Huawei Technologies Co., Ltd. 2026-2026. All rights reserved.
+ * Description: urma_ping parameters implementation file
+ * Author: Wang Hang
+ * Create: 2026-02-03
+ * Note:
+ * History: 2026-02-03 Create file
+ */
+
+#include <errno.h>
+#include <getopt.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "ping_log.h"
+
+#include "ping_parameters.h"
+
+int str_to_u32(const char *buf, uint32_t *u32)
+{
+    unsigned long ret;
+    char *end = NULL;
+
+    if (buf == NULL || *buf == '-') {
+        return -EINVAL;
+    }
+
+    errno = 0;
+    ret = strtoul(buf, &end, 0);
+    if (errno == ERANGE && ret == ULONG_MAX) {
+        return -EFAULT;
+    }
+    if (end == NULL || *end != '\0' || end == buf) {
+        return -ENOEXEC;
+    }
+    if (ret > UINT32_MAX) {
+        return -ERANGE;
+    }
+    *u32 = (uint32_t)ret;
+    return 0;
+}
+
+static inline void ipv4_map_to_eid(uint32_t ipv4, urma_eid_t *eid)
+{
+    const uint32_t ipv4_map_ipv6_prefix = 0x0000ffff;
+    eid->in4.reserved = 0;
+    eid->in4.prefix = htobe32(ipv4_map_ipv6_prefix);
+    eid->in4.addr = htobe32(ipv4);
+}
+
+int str_to_eid(const char *buf, urma_eid_t *eid)
+{
+    int ret;
+    uint32_t ipv4;
+    if (buf == NULL || eid == NULL) {
+        return -EINVAL;
+    }
+
+    // ipv6 addr
+    if (inet_pton(AF_INET6, buf, eid) > 0) {
+        return 0;
+    }
+
+    // ipv4 addr: xx.xx.xx.xx
+    if (inet_pton(AF_INET, buf, &ipv4) > 0) {
+        ipv4_map_to_eid(be32toh(ipv4), eid);
+        return 0;
+    }
+
+    // ipv4 value: 0x12345  or abcdef or 12345
+    ret = str_to_u32(buf, &ipv4);
+    if (ret == 0) {
+        ipv4_map_to_eid(ipv4, eid);
+        return 0;
+    }
+
+    return -EINVAL;
+}
+
+void version()
+{
+    LOG_QUIET("Version: 0.0.1\n");
+}
+
+void usage(const char *filename)
+{
+    LOG_QUIET("Usage:\n"
+              "  urma_ping [option] <destination>\n"
+              "\n"
+              "Options:\n"
+              "  <destination>    Primary eid or bonding eid\n"
+              "  -c <count>       Packet count (default: INT_MAX)\n"
+              "  -i <interval>    Interval between packets in seconds (default: 1)\n"
+              "  -s <size>        Send buffer size in bytes (default: 4)\n"
+              "  -w <deadline>    Total execution time limit in seconds (default: umlimited)\n"
+              "  -W <timeout>     Per-reply timeout in seconds (default: 1)\n"
+              "  -q               Quiet output\n"
+              "  -v --verbose     Verbose output\n"
+              "  -h --help        Show this help and exit\n"
+              "  -V --version     Show version and exit\n");
+}
+
+int parse_args(ping_cfg_t *cfg)
+{
+    static const struct option long_options[] = {
+        {"verbose", no_argument, NULL, 'v'},
+        {"help", no_argument, NULL, 'h'},
+        {"version", no_argument, NULL, 'V'},
+        {NULL, 0, NULL, 0},
+    };
+
+    int opt, ret = 0;
+    while ((opt = getopt_long(cfg->argc, cfg->argv, "c:i:s:qw:W:ShVv", long_options, NULL)) != -1) {
+        switch (opt) {
+            case 'c':
+                ret = str_to_u32(optarg, &cfg->count);
+                break;
+            case 'i':
+                ret = str_to_u32(optarg, &cfg->interval);
+                break;
+            case 's':
+                ret = str_to_u32(optarg, &cfg->size);
+                break;
+            case 'w':
+                ret = str_to_u32(optarg, &cfg->deadline);
+                break;
+            case 'W':
+                ret = str_to_u32(optarg, &cfg->timeout);
+                break;
+            case 'q':
+                cfg->verbose_level = VLOG_LEVEL_QUIET;
+                break;
+            case 'v':
+                cfg->verbose_level++;
+                break;
+            case 'h':
+                usage(cfg->filename);
+                exit(EXIT_SUCCESS);
+            case 'V':
+                version();
+                exit(EXIT_SUCCESS);
+            case ':':
+                LOG_ERROR("Option -%c requires an argument\n", optopt);
+                return -EINVAL;
+            default:
+                LOG_ERROR("Unknown option\n");
+                usage(cfg->filename);
+                return -EINVAL;
+        }
+        if (ret != 0) {
+            const int option_offset = 2;
+            LOG_ERROR("Invalid option %s\n", cfg->argv[optind - option_offset]);
+            return -EINVAL;
+        }
+    }
+
+    if (optind >= cfg->argc) {
+        LOG_ERROR("Destination is required\n");
+        usage(cfg->filename);
+        return -EINVAL;
+    }
+
+    ret = str_to_eid(cfg->argv[optind], &cfg->dest);
+    if (ret != 0) {
+        LOG_ERROR("Invalid dest eid %s\n", cfg->argv[optind]);
+        return -EINVAL;
+    }
+    return 0;
+}
+
+int check_args(ping_cfg_t *cfg)
+{
+    if (cfg->count == 0) {
+        LOG_ERROR("Count must be greater than 0\n");
+        return -EINVAL;
+    }
+
+    if (cfg->size == 0) {
+        LOG_ERROR("Size must be greater than 0\n");
+        return -EINVAL;
+    }
+
+    return 0;
+}
diff --git a/src/urma/tools/urma_ping/ping_parameters.h b/src/urma/tools/urma_ping/ping_parameters.h
new file mode 100644
index 0000000000000000000000000000000000000000..12335e1af7a318d323accabe249b6a0907fd95bf
--- /dev/null
+++ b/src/urma/tools/urma_ping/ping_parameters.h
@@ -0,0 +1,39 @@
+/*
+ * SPDX-License-Identifier: MIT
+ * Copyright (c) Huawei Technologies Co., Ltd. 2026-2026. All rights reserved.
+ * Description: urma_ping parameters head file
+ * Author: Wang Hang
+ * Create: 2026-02-03
+ * Note:
+ * History: 2026-02-03 Create file
+ */
+
+#ifndef URMA_PING_PARAMETERS_H
+#define URMA_PING_PARAMETERS_H
+
+#include <stdint.h>
+
+#include "urma_types.h"
+
+typedef struct ping_cfg {
+    int argc;
+    char **argv;
+    char *filename;
+    urma_eid_t dest;
+    uint32_t count;    /* UINT32_MAX means unlimited */
+    uint32_t interval; /* seconds */
+    uint32_t size;     /* bytes */
+    uint32_t deadline; /* seconds, 0 means disabled */
+    uint32_t timeout;  /* seconds, per-reply timeout */
+    uint32_t verbose_level;
+} ping_cfg_t;
+
+int str_to_u32(const char *buf, uint32_t *u32);
+int str_to_eid(const char *buf, urma_eid_t *eid);
+
+void version();
+void usage(const char *filename);
+int parse_args(ping_cfg_t *cfg);
+int check_args(ping_cfg_t *cfg);
+
+#endif
diff --git a/src/urma/tools/urma_ping/ping_run.c b/src/urma/tools/urma_ping/ping_run.c
new file mode 100644
index 0000000000000000000000000000000000000000..bd702abb1226b8c6cea57c20ad06f09cdaa3cb21
--- /dev/null
+++ b/src/urma/tools/urma_ping/ping_run.c
@@ -0,0 +1,615 @@
+/*
+ * SPDX-License-Identifier: MIT
+ * Copyright (c) Huawei Technologies Co., Ltd. 2026-2026. All rights reserved.
+ * Description: urma_ping run implementation file
+ * Author: Wang Hang
+ * Create: 2026-02-03
+ * Note:
+ * History: 2026-02-03 Create file
+ */
+
+#include <malloc.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "urma_api.h"
+#include "urma_types.h"
+
+#include "uvs_api.h"
+
+#include "ping_log.h"
+#include "ping_stat.h"
+
+#include "ping_run.h"
+
+const uint32_t PING_WK_JETTY_ID = 5;
+const uint32_t PING_SEND_DEPTH = 1024;
+const uint32_t PING_RECV_DEPTH = 1024;
+const uint32_t EID_STRLEN = 128;
+const uint32_t EID_MAX_VALUE = (1 << 30);
+
+typedef struct ping_urma_resource {
+    urma_context_t *ctx;
+    void *buf;
+    urma_target_seg_t *seg;
+    urma_jfc_t *send_jfc;
+    urma_jfc_t *recv_jfc;
+    urma_jfr_t *jfr;
+    urma_jetty_t *jetty;
+    urma_target_jetty_t *tjetty;
+} ping_urma_resource_t;
+
+static void log_eid_verbose(uint8_t *eid)
+{
+    char print_str[EID_STRLEN];
+    inet_ntop(AF_INET6, eid, print_str, EID_STRLEN);
+    LOG_VERBOSE("%s\n", print_str);
+}
+
+static void primary_eid_to_main_primary_eid(urma_eid_t *primary_eid)
+{
+    int ret;
+    urma_ping_ubcore_topo_map_t topo_map;
+
+    ret = uvs_get_topo_info(&topo_map);
+    if (ret != 0) {
+        LOG_ERROR("Failed to get ubagg topo, ret:%d\n", ret);
+        return;
+    }
+
+    bool find_entity_id = false;
+    uint32_t target_entity_id;
+    for (int i = 0; i < (int)topo_map.node_num; i++) {
+        for (int j = 0; j < DEV_NUM; j++) {
+            for (int k = 0; k < IODIE_NUM; k++) {
+                if (memcmp(topo_map.topo_infos[i].agg_devs[j].ues[k].primary_eid, primary_eid->raw, EID_LEN) == 0) {
+                    target_entity_id = topo_map.topo_infos[i].agg_devs[j].ues[k].entity_id;
+                    find_entity_id = true;
+                }
+            }
+        }
+    }
+
+    if (!find_entity_id) {
+        LOG_ERROR("Failed to get entity id\n");
+        return;
+    }
+
+    urma_eid_t empty_eid = {0};
+    for (int i = 0; i < (int)topo_map.node_num; i++) {
+        for (int j = 0; j < DEV_NUM; j++) {
+            for (int k = 0; k < IODIE_NUM; k++) {
+                if (topo_map.topo_infos[i].agg_devs[j].ues[k].entity_id == target_entity_id) {
+                    if (memcmp(topo_map.topo_infos[i].agg_devs[j].ues[k].primary_eid, empty_eid.raw, EID_LEN) == 0)
+                        continue;
+                    if (memcmp(primary_eid->raw, topo_map.topo_infos[i].agg_devs[j].ues[k].primary_eid, EID_LEN) > 0) {
+                        memcpy(primary_eid->raw, topo_map.topo_infos[i].agg_devs[j].ues[k].primary_eid, EID_LEN);
+                    }
+                }
+            }
+        }
+    }
+}
+
+static bool is_primary_eid(urma_eid_t *eid, urma_ping_ubcore_topo_map_t *topo_map)
+{
+    for (int i = 0; i < (int)topo_map->node_num; i++) {
+        for (int j = 0; j < DEV_NUM; j++) {
+            for (int k = 0; k < IODIE_NUM; k++) {
+                if (memcmp(eid->raw, topo_map->topo_infos[i].agg_devs[j].ues[k].primary_eid, EID_LEN) == 0) {
+                    return true;
+                }
+            }
+        }
+    }
+    return false;
+}
+
+static uint32_t get_chip_id_by_primary_eid(urma_eid_t *eid, urma_ping_ubcore_topo_map_t *topo_map)
+{
+    for (int i = 0; i < (int)topo_map->node_num; i++) {
+        for (int j = 0; j < DEV_NUM; j++) {
+            for (int k = 0; k < IODIE_NUM; k++) {
+                return topo_map->topo_infos[i].agg_devs[j].ues[k].chip_id;
+            }
+        }
+    }
+
+    LOG_ERROR("Failed to get chip id by primary eid.\n");
+    return 0;
+}
+
+static bool get_primary_eid_and_chip_id_by_bonding(urma_eid_t *original_eid, urma_eid_t *primary_eid, uint32_t *chip_id)
+{
+    int ret;
+    urma_ping_ubcore_topo_map_t topo_map;
+
+    ret = uvs_get_topo_info(&topo_map);
+    if (ret != 0) {
+        LOG_ERROR("Failed to get ubagg topo, ret:%d\n", ret);
+        return false;
+    }
+
+    if (is_primary_eid(original_eid, &topo_map)) {
+        memcpy(primary_eid->raw, original_eid->raw, EID_LEN);
+        (*chip_id) = get_chip_id_by_primary_eid(original_eid, &topo_map);
+        LOG_VERBOSE("Input dest eid is primary eid: ");
+        log_eid_verbose(primary_eid->raw);
+        LOG_VERBOSE("Chip id = %u.\n", (*chip_id));
+        return true;
+    }
+
+    LOG_VERBOSE("Dest Bonding eid: ");
+    log_eid_verbose(original_eid->raw);
+
+    for (int i = 0; i < (int)topo_map.node_num; i++) {
+        for (int j = 0; j < DEV_NUM; j++) {
+            if (memcmp(topo_map.topo_infos[i].agg_devs[j].agg_eid, original_eid->raw, EID_LEN) == 0) {
+                memcpy(primary_eid->raw, topo_map.topo_infos[i].agg_devs[j].ues[0].primary_eid, EID_LEN);
+                (*chip_id) = topo_map.topo_infos[i].agg_devs[j].ues[0].chip_id;
+                LOG_VERBOSE("Successful find dest primary eid: ");
+                log_eid_verbose(primary_eid->raw);
+                LOG_VERBOSE("Chip id = %u.\n", (*chip_id));
+                return true;
+            }
+        }
+    }
+
+    LOG_ERROR("Failed to find dest primary eid.\n");
+    return false;
+}
+
+static bool get_source_eid_by_bonding_eid_and_chip_id(urma_eid_t *eid, uint32_t chip_id)
+{
+    int ret;
+    urma_ping_ubcore_topo_map_t topo_map;
+
+    ret = uvs_get_topo_info(&topo_map);
+    if (ret != 0) {
+        LOG_ERROR("Failed to get ubagg topo, ret:%d\n", ret);
+        return false;
+    }
+
+    LOG_VERBOSE("Source Bonding eid: ");
+    log_eid_verbose(eid->raw);
+
+    for (int i = 0; i < (int)topo_map.node_num; i++) {
+        for (int j = 0; j < DEV_NUM; j++) {
+            if (memcmp(topo_map.topo_infos[i].agg_devs[j].agg_eid, eid->raw, EID_LEN) == 0) {
+                for (int k = 0; k < IODIE_NUM; k++) {
+                    if (topo_map.topo_infos[i].agg_devs[j].ues[k].chip_id == chip_id) {
+                        memcpy(eid->raw, topo_map.topo_infos[i].agg_devs[j].ues[k].primary_eid, EID_LEN);
+                        LOG_VERBOSE("Successful find source primary eid: ");
+                        log_eid_verbose(eid->raw);
+                        return true;
+                    }
+                }
+            }
+        }
+    }
+
+    LOG_ERROR("Failed to find source primary eid.\n");
+    return false;
+}
+
+static bool if_device_has_eid(urma_device_t *dev)
+{
+    uint32_t eid_cnt = 0;
+    urma_eid_info_t *eid_list = urma_get_eid_list(dev, &eid_cnt);
+    if (eid_list != NULL) {
+        urma_free_eid_list(eid_list);
+        return true;
+    }
+    return false;
+}
+
+static urma_device_t *find_first_urma_device()
+{
+    int num_devices = 0;
+    urma_device_t **devices = urma_get_device_list(&num_devices);
+    urma_device_t *device_selected = NULL;
+
+    for (int i = 0; i < num_devices; i++) {
+        if (device_selected == NULL && devices[i]->type == URMA_TRANSPORT_UB &&
+            strncmp(devices[i]->name, "bonding_dev_0", strlen("bonding_dev_0") + 1) == 0 &&
+            if_device_has_eid(devices[i])) {
+            device_selected = devices[i];
+        }
+    }
+
+    for (int i = 0; i < num_devices; i++) {
+        if (device_selected == NULL && devices[i]->type == URMA_TRANSPORT_UB &&
+            strncmp(devices[i]->name, "bonding_dev", strlen("bonding_dev")) == 0 &&
+            if_device_has_eid(devices[i])) {
+            device_selected = devices[i];
+        }
+    }
+
+    urma_free_device_list(devices);
+    return device_selected;
+}
+
+static void urma_log_func(int level, char *message)
+{
+    LOG_VVERBOSE("%s", message);
+}
+
+static int init_urma_resource(ping_cfg_t *cfg, ping_urma_resource_t *res)
+{
+    int ret = 0;
+    if ((ret = urma_register_log_func(urma_log_func)) != URMA_SUCCESS) {
+        LOG_ERROR("Failed to register urma log func, ret:%d\n", ret);
+        return ret;
+    }
+
+    urma_init_attr_t conf = {0};
+    if ((ret = urma_init(&conf)) != URMA_SUCCESS) {
+        LOG_ERROR("Failed to init urma, ret:%d\n", ret);
+        return ret;
+    }
+
+    urma_device_t *dev = find_first_urma_device();
+    if (dev == NULL) {
+        LOG_ERROR("Failed to find first urma device.\n");
+        ret = -ENODEV;
+        goto uninit_urma;
+    }
+
+    uint32_t eid_cnt = 1, chip_id = 0, source_eid_index = EID_MAX_VALUE;
+    urma_eid_t source_eid, dest_eid;
+    if (!get_primary_eid_and_chip_id_by_bonding(&(cfg->dest), &dest_eid, &chip_id)) {
+        LOG_ERROR("Failed to dest primary eid.\n");
+        ret = -ENODEV;
+        goto uninit_urma;
+    }
+
+    primary_eid_to_main_primary_eid(&dest_eid);
+
+    urma_eid_info_t *eid_list = urma_get_eid_list(dev, &eid_cnt);
+    if (eid_list == NULL) {
+        LOG_ERROR("Failed to get eid list.\n");
+        ret = -ENODEV;
+        goto uninit_urma;
+    }
+    if (eid_cnt != 1) {
+        LOG_ERROR("More than one eid for bonding_dev.\n");
+        ret = -ENODEV;
+        goto uninit_urma;
+    }
+
+    source_eid = eid_list[0].eid;
+    if (!get_source_eid_by_bonding_eid_and_chip_id(&source_eid, chip_id)) {
+        LOG_ERROR("Failed to source primary eid.\n");
+        ret = -ENODEV;
+        goto uninit_urma;
+    }
+
+    urma_device_t *primary_dev = urma_get_device_by_eid(source_eid, URMA_TRANSPORT_UB);
+    if (primary_dev == NULL) {
+        LOG_ERROR("Failed to find primary urma device.\n");
+        ret = -ENODEV;
+        goto uninit_urma;
+    }
+    LOG_VERBOSE("device = %s\n", primary_dev->name);
+
+    eid_list = urma_get_eid_list(primary_dev, &eid_cnt);
+    for (int i = 0; i < eid_cnt; i++) {
+        if (memcmp(&source_eid, &(eid_list[i].eid), sizeof(urma_eid_t)) == 0) {
+            source_eid_index = eid_list[i].eid_index;
+            break;
+        }
+    }
+    if (source_eid_index == EID_MAX_VALUE) {
+        LOG_ERROR("Failed to find primary eid_index.\n");
+        ret = -ENODEV;
+        goto uninit_urma;
+    }
+    LOG_VERBOSE("Get primary eid_index = %d\n", (int)source_eid_index);
+
+    urma_device_attr_t dev_attr = {0};
+    if ((ret = urma_query_device(primary_dev, &dev_attr)) != URMA_SUCCESS) {
+        LOG_ERROR("Failed to query urma device attr, ret:%d\n", ret);
+        goto uninit_urma;
+    }
+    if (cfg->size > dev_attr.dev_cap.max_msg_size) {
+        LOG_ERROR("Ping message size %u exceeds device max msg size %u.\n",
+                  cfg->size, dev_attr.dev_cap.max_msg_size);
+        ret = -EINVAL;
+        goto uninit_urma;
+    }
+
+    res->ctx = urma_create_context(primary_dev, source_eid_index);
+    if (res->ctx == NULL) {
+        LOG_ERROR("Failed to create urma context.\n");
+        ret = -EINVAL;
+        goto uninit_urma;
+    }
+
+    res->buf = memalign(getpagesize(), cfg->size);
+    if (res->buf == NULL) {
+        LOG_ERROR("Failed to alloc memory for ping buffer.\n");
+        ret = -ENOMEM;
+        goto delete_context;
+    }
+
+    urma_seg_cfg_t seg_cfg = {
+        .va = (uint64_t)(uintptr_t)res->buf,
+        .len = cfg->size,
+        .flag.bs.access = URMA_ACCESS_LOCAL_ONLY,
+    };
+    res->seg = urma_register_seg(res->ctx, &seg_cfg);
+    if (res->seg == NULL) {
+        LOG_ERROR("Failed to register urma segment.\n");
+        ret = -EINVAL;
+        goto free_buf;
+    }
+
+    urma_jfc_cfg_t send_jfc_cfg = {
+        .depth = PING_SEND_DEPTH,
+    };
+    res->send_jfc = urma_create_jfc(res->ctx, &send_jfc_cfg);
+    if (res->send_jfc == NULL) {
+        LOG_ERROR("Failed to create urma jfc for send.\n");
+        ret = -EINVAL;
+        goto unregister_seg;
+    }
+
+    urma_jfc_cfg_t recv_jfc_cfg = {
+        .depth = PING_RECV_DEPTH,
+    };
+    res->recv_jfc = urma_create_jfc(res->ctx, &recv_jfc_cfg);
+    if (res->recv_jfc == NULL) {
+        LOG_ERROR("Failed to create urma jfc for recv.\n");
+        ret = -EINVAL;
+        goto delete_send_jfc;
+    }
+
+    urma_jfr_cfg_t jfr_cfg = {
+        .depth = PING_RECV_DEPTH,
+        .trans_mode = URMA_TM_RM,
+        .max_sge = 1,
+        .jfc = res->recv_jfc,
+    };
+    res->jfr = urma_create_jfr(res->ctx, &jfr_cfg);
+    if (res->jfr == NULL) {
+        LOG_ERROR("Failed to create urma jfr.\n");
+        ret = -EINVAL;
+        goto delete_recv_jfc;
+    }
+
+    urma_jetty_cfg_t jetty_cfg = {
+        .flag.bs.share_jfr = 1,
+        .jfs_cfg = {
+            .depth = PING_SEND_DEPTH,
+            .trans_mode = URMA_TM_RM,
+            .priority = 6,
+            .max_sge = 1,
+            .rnr_retry = URMA_TYPICAL_RNR_RETRY,
+            .err_timeout = URMA_TYPICAL_ERR_TIMEOUT,
+            .jfc = res->send_jfc,
+        },
+        .shared.jfr = res->jfr,
+    };
+    res->jetty = urma_create_jetty(res->ctx, &jetty_cfg);
+    if (res->jetty == NULL) {
+        LOG_ERROR("Failed to create urma jetty.\n");
+        ret = -EINVAL;
+        goto delete_jfr;
+    }
+
+    urma_rjetty_t rjetty = {
+        .jetty_id.eid = dest_eid,
+        .jetty_id.id = PING_WK_JETTY_ID,
+        .trans_mode = URMA_TM_RM,
+        .type = URMA_JETTY,
+        .tp_type = URMA_CTP,
+    };
+    urma_token_t token_value = {0};
+    res->tjetty = urma_import_jetty(res->ctx, &rjetty, &token_value);
+    if (res->tjetty == NULL) {
+        LOG_ERROR("Failed to import target urma jetty.\n");
+        ret = -EINVAL;
+        goto delete_jetty;
+    }
+
+    return 0;
+
+delete_jetty:
+    urma_delete_jetty(res->jetty);
+delete_jfr:
+    urma_delete_jfr(res->jfr);
+delete_recv_jfc:
+    urma_delete_jfc(res->recv_jfc);
+delete_send_jfc:
+    urma_delete_jfc(res->send_jfc);
+unregister_seg:
+    urma_unregister_seg(res->seg);
+free_buf:
+    free(res->buf);
+delete_context:
+    urma_delete_context(res->ctx);
+uninit_urma:
+    urma_uninit();
+    return ret;
+}
+
+static void uninit_urma_resource(ping_urma_resource_t *res)
+{
+    urma_unimport_jetty(res->tjetty);
+    urma_delete_jetty(res->jetty);
+    urma_delete_jfr(res->jfr);
+    urma_unregister_seg(res->seg);
+    free(res->buf);
+    urma_delete_context(res->ctx);
+    urma_uninit();
+}
+
+typedef struct ping_per_seq_info {
+    uint32_t seq;
+    clock_t start_time;
+    clock_t end_time;
+    double rtt;
+} ping_per_seq_info_t;
+
+static int fill_recv_wr(ping_cfg_t *cfg, ping_urma_resource_t *res)
+{
+    urma_sge_t sge = {
+        .addr = (uint64_t)(uintptr_t)res->buf,
+        .len = cfg->size,
+        .tseg = res->seg,
+    };
+
+    urma_jfr_wr_t wr = {
+        .src = {
+            .sge = &sge,
+            .num_sge = 1,
+        },
+    };
+    urma_jfr_wr_t *bad_wr_placeholder = NULL;
+    return urma_post_jetty_recv_wr(res->jetty, &wr, &bad_wr_placeholder);
+}
+
+static int send_ping_msg(ping_cfg_t *cfg, ping_urma_resource_t *res, ping_per_seq_info_t *seq_info)
+{
+    update_stat_on_send();
+    seq_info->start_time = clock();
+
+    urma_sge_t sge = {
+        .addr = (uint64_t)(uintptr_t)res->buf,
+        .len = cfg->size,
+        .tseg = res->seg,
+    };
+    urma_jfs_wr_t wr = {
+        .opcode = URMA_OPC_SEND_IMM,
+        .flag.bs.complete_enable = 0,
+        .tjetty = res->tjetty,
+        .send = {
+            .src = {
+                .sge = &sge,
+                .num_sge = 1,
+            },
+            .imm_data = seq_info->seq,
+        },
+    };
+    urma_jfs_wr_t *badwr = NULL;
+    int ret = urma_post_jetty_send_wr(res->jetty, &wr, &badwr);
+    if (ret != 0) {
+        LOG_ERROR("Failed to post jetty send wr, ret:%d\n", ret);
+        return ret;
+    }
+
+    return 0;
+}
+
+static int recv_ping_msg(ping_cfg_t *cfg, ping_urma_resource_t *res, ping_per_seq_info_t *seq_info)
+{
+    bool timed_out = false;
+    urma_cr_t cr = {0};
+    while (true) {
+        int ret = urma_poll_jfc(res->recv_jfc, 1, &cr);
+
+        seq_info->end_time = clock();
+        seq_info->rtt = (double)(seq_info->end_time - seq_info->start_time) * 1000 / CLOCKS_PER_SEC;
+
+        bool unmatched = (ret > 0 && cr.imm_data != seq_info->seq);
+        bool received = (ret > 0 && cr.imm_data == seq_info->seq);
+        timed_out = (seq_info->rtt >= cfg->timeout * 1000);
+
+        if (unmatched) {
+            LOG_VERBOSE("Unmatched reply received: expected seq=%d, got seq=%d\n",
+                        seq_info->seq, cr.imm_data);
+        }
+        if (received || timed_out) {
+            break;
+        }
+    }
+
+    if (timed_out) {
+        // timeout
+        LOG_NORMAL("Request timeout for seq=%d\n", seq_info->seq);
+    } else {
+        int ret = fill_recv_wr(cfg, res);
+        if (ret != 0) {
+            LOG_ERROR("Failed to fill recv wr, ret:%d\n", ret);
+            return ret;
+        }
+
+        if (cr.status == URMA_CR_SUCCESS) {
+            // success
+            update_stat_on_recv(seq_info->rtt);
+            LOG_NORMAL("%d bytes from " EID_FMT ": seq=%d time=%.3f ms\n",
+                       cfg->size, EID_ARGS(cfg->dest), seq_info->seq, seq_info->rtt);
+        } else {
+            // failure
+            LOG_NORMAL("From " EID_FMT " seq=%d cr_status=%d\n",
+                       EID_ARGS(cfg->dest), seq_info->seq, cr.status);
+        }
+    }
+
+    return 0;
+}
+
+int start_ping(ping_cfg_t *cfg)
+{
+    verbose_set_level(cfg->verbose_level);
+
+    int ret = 0;
+    ping_urma_resource_t res = {0};
+
+    if ((ret = init_urma_resource(cfg, &res)) != 0) {
+        return ret;
+    }
+
+    if ((ret = fill_recv_wr(cfg, &res)) != 0) {
+        LOG_ERROR("Failed to prepare recv wr, ret:%d\n", ret);
+        return ret;
+    }
+
+    LOG_QUIET("URMA_PING " EID_FMT " %u bytes of data.\n", EID_ARGS(cfg->dest), cfg->size);
+    LOG_VERBOSE("Count       : %u\n", cfg->count);
+    LOG_VERBOSE("Interval(s) : %u\n", cfg->interval);
+    LOG_VERBOSE("Size(bytes) : %u\n", cfg->size);
+    LOG_VERBOSE("Deadline(s) : %u\n", cfg->deadline);
+    LOG_VERBOSE("Timeout(s)  : %u\n", cfg->timeout);
+
+    init_stat();
+
+    uint32_t ping_start_time = clock();
+
+    for (uint32_t i = 0; i < cfg->count; i++) {
+        ping_per_seq_info_t seq_info = {
+            .seq = i + 1,
+        };
+
+        ret = send_ping_msg(cfg, &res, &seq_info);
+        if (ret != 0) {
+            LOG_ERROR("Failed to send ping msg, ret:%d\n", ret);
+            break;
+        }
+
+        ret = recv_ping_msg(cfg, &res, &seq_info);
+        if (ret != 0) {
+            LOG_ERROR("Failed to receive ping msg, ret:%d\n", ret);
+            break;
+        }
+
+        if (i + 1 < cfg->count) {
+            sleep(cfg->interval);
+        }
+
+        uint32_t time_elapsed = (clock() - ping_start_time) / CLOCKS_PER_SEC;
+        LOG_VERBOSE("Time elapsed = %d\n", (int)time_elapsed);
+        if (cfg->deadline != 0 && time_elapsed > cfg->deadline) {
+            LOG_ERROR("URMA_PING time limit exceeded.\n");
+            break;
+        }
+    }
+
+    if (false) {
+        uninit_urma_resource(&res);
+    }
+    return ret;
+}
diff --git a/src/urma/tools/urma_ping/ping_run.h b/src/urma/tools/urma_ping/ping_run.h
new file mode 100644
index 0000000000000000000000000000000000000000..9e813d816790801587830a7e71b735e877fd1455
--- /dev/null
+++ b/src/urma/tools/urma_ping/ping_run.h
@@ -0,0 +1,56 @@
+/*
+ * SPDX-License-Identifier: MIT
+ * Copyright (c) Huawei Technologies Co., Ltd. 2026-2026. All rights reserved.
+ * Description: urma_ping run head file
+ * Author: Wang Hang
+ * Create: 2026-02-03
+ * Note:
+ * History: 2026-02-03 Create file
+ */
+
+#ifndef URMA_PING_RUN_H
+#define URMA_PING_RUN_H
+
+#include "ping_parameters.h"
+
+#define EID_LEN            (16)
+#define MAX_PORT_NUM       (9)
+#define MAX_NODE_NUM       (16)
+#define IODIE_NUM          (2)
+#define PORT_NUM           (9)
+#define DEV_NUM            (128)
+#define ENTITY_AGG_DEV_NUM (3) // bonding device number per entity
+
+struct urma_ping_ubcore_topo_ue {
+    uint32_t chip_id;
+    uint32_t entity_id;
+    char primary_eid[EID_LEN];
+    char port_eid[PORT_NUM][EID_LEN];
+};
+
+struct urma_ping_ubcore_topo_agg_dev {
+    char agg_eid[EID_LEN];
+    struct urma_ping_ubcore_topo_ue ues[IODIE_NUM];
+};
+
+struct urma_ping_ubcore_topo_link {
+    uint32_t peer_node;  // node id
+    uint32_t peer_iodie; // iodie idx
+    uint32_t peer_port;  // port idx, UINT32_MAX indicates no connection
+};
+
+struct urma_ping_ubcore_topo_node {
+    uint32_t id;
+    uint32_t is_current;
+    struct urma_ping_ubcore_topo_link links[IODIE_NUM][PORT_NUM];
+    struct urma_ping_ubcore_topo_agg_dev agg_devs[DEV_NUM];
+};
+
+typedef struct urma_ping_ubcore_topo_map {
+    struct urma_ping_ubcore_topo_node topo_infos[MAX_NODE_NUM];
+    uint32_t node_num;
+} urma_ping_ubcore_topo_map_t;
+
+int start_ping(ping_cfg_t *cfg);
+
+#endif
diff --git a/src/urma/tools/urma_ping/ping_stat.c b/src/urma/tools/urma_ping/ping_stat.c
new file mode 100644
index 0000000000000000000000000000000000000000..a2eb6e9fca3d9cca133e165f63479870429da251
--- /dev/null
+++ b/src/urma/tools/urma_ping/ping_stat.c
@@ -0,0 +1,70 @@
+/*
+ * SPDX-License-Identifier: MIT
+ * Copyright (c) Huawei Technologies Co., Ltd. 2026-2026. All rights reserved.
+ * Description: urma_ping stat implementation file
+ * Author: Wang Hang
+ * Create: 2026-02-03
+ * Note:
+ * History: 2026-02-03 Create file
+ */
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <time.h>
+
+#include "ping_log.h"
+
+#include "ping_stat.h"
+
+typedef struct ping_stat {
+    uint32_t transmitted;
+    uint32_t received;
+    double rtt_min;
+    double rtt_max;
+    double rtt_avg;
+    double rtt_m2;
+    clock_t start_time;
+    clock_t end_time;
+} ping_stat_t;
+
+static ping_stat_t g_stat = {0};
+
+static void print_stat(void)
+{
+    g_stat.end_time = clock();
+
+    double total_time = (double)(g_stat.end_time - g_stat.start_time) * 1000 / CLOCKS_PER_SEC;
+    double loss_rate = ((g_stat.transmitted - g_stat.received) * 100.0) / g_stat.transmitted;
+    double rtt_mdev = g_stat.received > 1 ? g_stat.rtt_m2 / (g_stat.received - 1) : 0;
+
+    LOG_QUIET("\n");
+    LOG_QUIET("--- ping statistics ---\n");
+    LOG_QUIET("%u packets transmitted, %u received, %.1f%% packet loss, time %.0fms\n", g_stat.transmitted,
+              g_stat.received, loss_rate, total_time);
+    LOG_QUIET("rtt min/avg/max/mdev = %.3f/%.3f/%.3f/%.3f ms\n", g_stat.rtt_min, g_stat.rtt_avg, g_stat.rtt_max,
+              rtt_mdev);
+}
+
+void init_stat(void)
+{
+    atexit(print_stat);
+    g_stat.start_time = clock();
+}
+
+void update_stat_on_send(void)
+{
+    g_stat.transmitted++;
+}
+
+void update_stat_on_recv(double rtt)
+{
+    if (rtt < g_stat.rtt_min || g_stat.received == 0) {
+        g_stat.rtt_min = rtt;
+    }
+    if (rtt > g_stat.rtt_max) {
+        g_stat.rtt_max = rtt;
+    }
+    g_stat.rtt_avg = ((g_stat.rtt_avg * g_stat.received) + rtt) / (g_stat.received + 1);
+    g_stat.rtt_m2 += (rtt - g_stat.rtt_avg) * (rtt - g_stat.rtt_avg);
+    g_stat.received++;
+}
diff --git a/src/urma/tools/urma_ping/ping_stat.h b/src/urma/tools/urma_ping/ping_stat.h
new file mode 100644
index 0000000000000000000000000000000000000000..c6fc354416c2788fb406ef6ac215aaf79f8265cd
--- /dev/null
+++ b/src/urma/tools/urma_ping/ping_stat.h
@@ -0,0 +1,18 @@
+/*
+ * SPDX-License-Identifier: MIT
+ * Copyright (c) Huawei Technologies Co., Ltd. 2026-2026. All rights reserved.
+ * Description: urma_ping stat head file
+ * Author: Wang Hang
+ * Create: 2026-02-03
+ * Note:
+ * History: 2026-02-03 Create file
+ */
+
+#ifndef URMA_PING_STAT_H
+#define URMA_PING_STAT_H
+
+void init_stat(void);
+void update_stat_on_send(void);
+void update_stat_on_recv(double rtt);
+
+#endif
diff --git a/umdk.spec b/umdk.spec
index 1f3733a4bdbbffdf8c0b61f1174a25a7a923451a..45dc9bcd40f135c634df57274eeb70cbd46c692d 100644
--- a/umdk.spec
+++ b/umdk.spec
@@ -349,6 +349,7 @@ fi
 %files urma-tools
 %defattr(-,root,root)
     %{_bindir}/urma_admin
+    %{_bindir}/urma_ping
     /etc/rsyslog.d/urma_admin.conf
     %{_bindir}/urma_perftest
     %{_bindir}/ubagg_cli
