diff --git a/src/urma/tools/urma_ping/CMakeLists.txt b/src/urma/tools/urma_ping/CMakeLists.txt
index 4c34432a18e5975a3cc50ae528de897497253d4a..7d73ec45da771bc9e2e4bd5def96a7db457033e8 100644
--- a/src/urma/tools/urma_ping/CMakeLists.txt
+++ b/src/urma/tools/urma_ping/CMakeLists.txt
@@ -23,6 +23,7 @@ target_link_libraries(urma_ping
     PRIVATE
         urma
         tpsa
+        m
 )
 
 install(TARGETS urma_ping
diff --git a/src/urma/tools/urma_ping/ping_run.c b/src/urma/tools/urma_ping/ping_run.c
index bd702abb1226b8c6cea57c20ad06f09cdaa3cb21..fa8aaa3e73e4fbf942412383231208cf21944159 100644
--- a/src/urma/tools/urma_ping/ping_run.c
+++ b/src/urma/tools/urma_ping/ping_run.c
@@ -62,11 +62,14 @@ static void primary_eid_to_main_primary_eid(urma_eid_t *primary_eid)
 
     bool find_entity_id = false;
     uint32_t target_entity_id;
+    int target_node_id;
+    
     for (int i = 0; i < (int)topo_map.node_num; i++) {
         for (int j = 0; j < DEV_NUM; j++) {
             for (int k = 0; k < IODIE_NUM; k++) {
                 if (memcmp(topo_map.topo_infos[i].agg_devs[j].ues[k].primary_eid, primary_eid->raw, EID_LEN) == 0) {
                     target_entity_id = topo_map.topo_infos[i].agg_devs[j].ues[k].entity_id;
+                    target_node_id = i;
                     find_entity_id = true;
                 }
             }
@@ -79,15 +82,14 @@ static void primary_eid_to_main_primary_eid(urma_eid_t *primary_eid)
     }
 
     urma_eid_t empty_eid = {0};
-    for (int i = 0; i < (int)topo_map.node_num; i++) {
-        for (int j = 0; j < DEV_NUM; j++) {
-            for (int k = 0; k < IODIE_NUM; k++) {
-                if (topo_map.topo_infos[i].agg_devs[j].ues[k].entity_id == target_entity_id) {
-                    if (memcmp(topo_map.topo_infos[i].agg_devs[j].ues[k].primary_eid, empty_eid.raw, EID_LEN) == 0)
-                        continue;
-                    if (memcmp(primary_eid->raw, topo_map.topo_infos[i].agg_devs[j].ues[k].primary_eid, EID_LEN) > 0) {
-                        memcpy(primary_eid->raw, topo_map.topo_infos[i].agg_devs[j].ues[k].primary_eid, EID_LEN);
-                    }
+
+    for (int j = 0; j < DEV_NUM; j++) {
+        for (int k = 0; k < IODIE_NUM; k++) {
+            if (topo_map.topo_infos[target_node_id].agg_devs[j].ues[k].entity_id == target_entity_id) {
+                if (memcmp(topo_map.topo_infos[target_node_id].agg_devs[j].ues[k].primary_eid, empty_eid.raw, EID_LEN) == 0)
+                    continue;
+                if (memcmp(primary_eid->raw, topo_map.topo_infos[target_node_id].agg_devs[j].ues[k].primary_eid, EID_LEN) > 0) {
+                    memcpy(primary_eid->raw, topo_map.topo_infos[target_node_id].agg_devs[j].ues[k].primary_eid, EID_LEN);
                 }
             }
         }
@@ -268,6 +270,9 @@ static int init_urma_resource(ping_cfg_t *cfg, ping_urma_resource_t *res)
 
     primary_eid_to_main_primary_eid(&dest_eid);
 
+    LOG_VERBOSE("Successful find dest primary eid: ");
+    log_eid_verbose(dest_eid.raw);
+
     urma_eid_info_t *eid_list = urma_get_eid_list(dev, &eid_cnt);
     if (eid_list == NULL) {
         LOG_ERROR("Failed to get eid list.\n");
@@ -449,8 +454,8 @@ static void uninit_urma_resource(ping_urma_resource_t *res)
 
 typedef struct ping_per_seq_info {
     uint32_t seq;
-    clock_t start_time;
-    clock_t end_time;
+    double start_time;
+    double end_time;
     double rtt;
 } ping_per_seq_info_t;
 
@@ -475,7 +480,7 @@ static int fill_recv_wr(ping_cfg_t *cfg, ping_urma_resource_t *res)
 static int send_ping_msg(ping_cfg_t *cfg, ping_urma_resource_t *res, ping_per_seq_info_t *seq_info)
 {
     update_stat_on_send();
-    seq_info->start_time = clock();
+    seq_info->start_time = get_time_in_ms();
 
     urma_sge_t sge = {
         .addr = (uint64_t)(uintptr_t)res->buf,
@@ -511,8 +516,8 @@ static int recv_ping_msg(ping_cfg_t *cfg, ping_urma_resource_t *res, ping_per_se
     while (true) {
         int ret = urma_poll_jfc(res->recv_jfc, 1, &cr);
 
-        seq_info->end_time = clock();
-        seq_info->rtt = (double)(seq_info->end_time - seq_info->start_time) * 1000 / CLOCKS_PER_SEC;
+        seq_info->end_time = get_time_in_ms();
+        seq_info->rtt = seq_info->end_time - seq_info->start_time;
 
         bool unmatched = (ret > 0 && cr.imm_data != seq_info->seq);
         bool received = (ret > 0 && cr.imm_data == seq_info->seq);
@@ -577,7 +582,7 @@ int start_ping(ping_cfg_t *cfg)
 
     init_stat();
 
-    uint32_t ping_start_time = clock();
+    double ping_start_time = get_time_in_ms();
 
     for (uint32_t i = 0; i < cfg->count; i++) {
         ping_per_seq_info_t seq_info = {
@@ -600,9 +605,10 @@ int start_ping(ping_cfg_t *cfg)
             sleep(cfg->interval);
         }
 
-        uint32_t time_elapsed = (clock() - ping_start_time) / CLOCKS_PER_SEC;
-        LOG_VERBOSE("Time elapsed = %d\n", (int)time_elapsed);
-        if (cfg->deadline != 0 && time_elapsed > cfg->deadline) {
+        double time_elapsed = get_time_in_ms() - ping_start_time;
+
+        LOG_VERBOSE("Time elapsed = %.3lfms\n", time_elapsed);
+        if (cfg->deadline != 0 && time_elapsed > cfg->deadline * 1000.0) {
             LOG_ERROR("URMA_PING time limit exceeded.\n");
             break;
         }
diff --git a/src/urma/tools/urma_ping/ping_stat.c b/src/urma/tools/urma_ping/ping_stat.c
index a2eb6e9fca3d9cca133e165f63479870429da251..f789e58ca409f39b4ea25ae1f1dd691e69bbe4c2 100644
--- a/src/urma/tools/urma_ping/ping_stat.c
+++ b/src/urma/tools/urma_ping/ping_stat.c
@@ -11,6 +11,7 @@
 #include <stdint.h>
 #include <stdlib.h>
 #include <time.h>
+#include <math.h>
 
 #include "ping_log.h"
 
@@ -23,19 +24,30 @@ typedef struct ping_stat {
     double rtt_max;
     double rtt_avg;
     double rtt_m2;
-    clock_t start_time;
-    clock_t end_time;
+    double start_time;
+    double end_time;
 } ping_stat_t;
 
 static ping_stat_t g_stat = {0};
 
+double get_time_in_ms(void)
+{
+    struct timespec cur;
+
+    if (clock_gettime(CLOCK_MONOTONIC, &cur) == -1) {
+        LOG_ERROR("clock_gettime failed.\n");
+        return 0;
+    }
+    return cur.tv_sec * 1000.0 + cur.tv_nsec / 1000000.0;
+}
+
 static void print_stat(void)
 {
-    g_stat.end_time = clock();
+    g_stat.end_time = get_time_in_ms();
 
-    double total_time = (double)(g_stat.end_time - g_stat.start_time) * 1000 / CLOCKS_PER_SEC;
+    double total_time = g_stat.end_time - g_stat.start_time;
     double loss_rate = ((g_stat.transmitted - g_stat.received) * 100.0) / g_stat.transmitted;
-    double rtt_mdev = g_stat.received > 1 ? g_stat.rtt_m2 / (g_stat.received - 1) : 0;
+    double rtt_mdev = g_stat.received > 0 ? sqrt(g_stat.rtt_m2 / g_stat.received) : 0;
 
     LOG_QUIET("\n");
     LOG_QUIET("--- ping statistics ---\n");
@@ -48,7 +60,7 @@ static void print_stat(void)
 void init_stat(void)
 {
     atexit(print_stat);
-    g_stat.start_time = clock();
+    g_stat.start_time = get_time_in_ms();
 }
 
 void update_stat_on_send(void)
@@ -64,7 +76,10 @@ void update_stat_on_recv(double rtt)
     if (rtt > g_stat.rtt_max) {
         g_stat.rtt_max = rtt;
     }
+
+    double old_rtt_avg = g_stat.rtt_avg;
+
     g_stat.rtt_avg = ((g_stat.rtt_avg * g_stat.received) + rtt) / (g_stat.received + 1);
-    g_stat.rtt_m2 += (rtt - g_stat.rtt_avg) * (rtt - g_stat.rtt_avg);
+    g_stat.rtt_m2 += (rtt - old_rtt_avg) * (rtt - g_stat.rtt_avg);
     g_stat.received++;
 }
diff --git a/src/urma/tools/urma_ping/ping_stat.h b/src/urma/tools/urma_ping/ping_stat.h
index c6fc354416c2788fb406ef6ac215aaf79f8265cd..85418b6bd97a1522f93c0fb8063aea9b762f3fdc 100644
--- a/src/urma/tools/urma_ping/ping_stat.h
+++ b/src/urma/tools/urma_ping/ping_stat.h
@@ -11,6 +11,7 @@
 #ifndef URMA_PING_STAT_H
 #define URMA_PING_STAT_H
 
+double get_time_in_ms(void);
 void init_stat(void);
 void update_stat_on_send(void);
 void update_stat_on_recv(double rtt);
