diff --git a/src/urma/lib/urma/bond/bondp_bitmap.c b/src/urma/lib/urma/bond/bondp_bitmap.c
deleted file mode 100644
index a95115513cb5334ec48d987e3f2256ee7ad210fd..0000000000000000000000000000000000000000
--- a/src/urma/lib/urma/bond/bondp_bitmap.c
+++ /dev/null
@@ -1,134 +0,0 @@
-/*
- * SPDX-License-Identifier: MIT
- * Copyright (c) Huawei Technologies Co., Ltd. 2025-2025. All rights reserved.
- * Description: Bond Bitmap implementation
- * Author: Ma Chuan
- * Create: 2025-02-18
- * Note:
- * History: 2025-02-18   Create File
- */
-#include <stdlib.h>
-#include <limits.h>
-#include "ub_bitmap.h"
-#include "ub_util.h"
-#include "urma_log.h"
-#include "bondp_bitmap.h"
-
-bondp_bitmap_t *bondp_bitmap_alloc(uint32_t bitmap_size)
-{
-    bondp_bitmap_t *bitmap = calloc(1, sizeof(bondp_bitmap_t));
-    if (bitmap == NULL) {
-        return NULL;
-    }
-    bitmap->size = bitmap_size;
-    bitmap->bits = ub_bitmap_alloc(bitmap_size);
-    if (bitmap->bits == NULL) {
-        free(bitmap);
-        return NULL;
-    }
-    return bitmap;
-}
-
-void bondp_bitmap_free(bondp_bitmap_t *bitmap)
-{
-    if (bitmap->bits != NULL) {
-        ub_bitmap_free(bitmap->bits);
-    }
-    free(bitmap);
-}
-
-int bondp_bitmap_init(bondp_bitmap_t *bitmap, uint32_t size)
-{
-    bitmap->size = size;
-    bitmap->bits = ub_bitmap_alloc(size);
-    if (bitmap->bits == NULL) {
-        return -1;
-    }
-    return 0;
-}
-
-void bondp_bitmap_uninit(bondp_bitmap_t *bitmap)
-{
-    if (bitmap->bits != NULL) {
-        ub_bitmap_free(bitmap->bits);
-    }
-}
-
-static unsigned long bondp_find_first_zero_bit(const unsigned long *array, unsigned long size)
-{
-    unsigned long i;
-    unsigned long ret = 0;
-    unsigned long array_size;
-
-    if (UB_UNLIKELY(size == 0)) {
-        return size;
-    }
-
-    array_size = ub_bitmap_n_ulongs(size);
-    for (i = 0; i < array_size; i++) {
-        if (array[i] == ULONG_MAX) {
-            ret += UB_ULONG_BITS;
-        } else {
-            return (ret + ffz(array[i]));
-        }
-    }
-
-    return size;
-}
-
-int bondp_bitmap_alloc_idx(bondp_bitmap_t *bitmap, uint32_t *idx_out)
-{
-    uint32_t idx;
-    idx = (uint32_t)bondp_find_first_zero_bit(bitmap->bits, bitmap->size);
-    if (idx >= bitmap->size) {
-        return -1;
-    }
-    ub_bitmap_set1(bitmap->bits, idx);
-    *idx_out = idx;
-    return 0;
-}
-
-int bondp_bitmap_alloc_idx_from_offset(bondp_bitmap_t *bitmap, uint32_t offset, uint32_t *idx_out)
-{
-    uint32_t idx = 0;
-    uint32_t offset_n_ulongs = 0;
-
-    if (offset >= bitmap->size) {
-        URMA_LOG_ERR("offset(%u) exceeds the bitmap size(%u).\n", offset, bitmap->size);
-        return -1;
-    }
-
-    offset_n_ulongs = ub_bitmap_n_ulongs(offset);
-    idx = (uint32_t)bondp_find_first_zero_bit(bitmap->bits + offset_n_ulongs, bitmap->size - offset);
-    if (idx >= bitmap->size - offset_n_ulongs) {
-        URMA_LOG_ERR("bitmap allocation failed.\n");
-        return -1;
-    }
-    ub_bitmap_set1(bitmap->bits + offset_n_ulongs, idx);
-    *idx_out = idx + offset_n_ulongs * sizeof(unsigned long) * CHAR_BIT;
-    return 0;
-}
-
-int bondp_bitmap_use_id(bondp_bitmap_t *bitmap, uint32_t id)
-{
-    if (ub_bitmap_is_set(bitmap->bits, id)) {
-        URMA_LOG_ERR("Bit %u is already taken.\n", id);
-        return -1;
-    }
-    ub_bitmap_set1(bitmap->bits, id);
-    return 0;
-}
-
-int bondp_bitmap_free_idx(bondp_bitmap_t *bitmap, uint32_t idx)
-{
-    if (!ub_bitmap_is_set(bitmap->bits, idx)) {
-        return -EINVAL;
-    }
-    ub_bitmap_set0(bitmap->bits, idx);
-    return 0;
-}
-
-bool bondp_bitmap_is_set(bondp_bitmap_t *bitmap, uint32_t id)
-{
-    return ub_bitmap_is_set(bitmap->bits, id);
-}
\ No newline at end of file
diff --git a/src/urma/lib/urma/bond/bondp_bitmap.h b/src/urma/lib/urma/bond/bondp_bitmap.h
deleted file mode 100644
index de776add8f49b25ad8f8fe44f36e594646c0ea97..0000000000000000000000000000000000000000
--- a/src/urma/lib/urma/bond/bondp_bitmap.h
+++ /dev/null
@@ -1,29 +0,0 @@
-/*
- * SPDX-License-Identifier: MIT
- * Copyright (c) Huawei Technologies Co., Ltd. 2025-2025. All rights reserved.
- * Description: Bond Bitmap header
- * Author: Ma Chuan
- * Create: 2025-02-18
- * Note:
- * History: 2025-02-18   Create File
- */
-#ifndef BONDP_BITMAP_H
-#define BONDP_BITMAP_H
-#include <stdbool.h>
-#include <inttypes.h>
-// Currently not multi-thread safe
-typedef struct bondp_bitmap {
-    unsigned long *bits;
-    uint32_t size;
-} bondp_bitmap_t;
-// -- All apis of bondp_bitmap need `Caller` to check non-null bitmap pointer --
-bondp_bitmap_t *bondp_bitmap_alloc(uint32_t bitmap_size);
-void bondp_bitmap_free(bondp_bitmap_t *bitmap);
-int bondp_bitmap_init(bondp_bitmap_t *bitmap, uint32_t size);
-void bondp_bitmap_uninit(bondp_bitmap_t *bitmap);
-int bondp_bitmap_alloc_idx(bondp_bitmap_t *bitmap, uint32_t *idx_out);
-int bondp_bitmap_alloc_idx_from_offset(bondp_bitmap_t *bitmap, uint32_t offset, uint32_t *idx_out);
-int bondp_bitmap_free_idx(bondp_bitmap_t *bitmap, uint32_t idx);
-int bondp_bitmap_use_id(bondp_bitmap_t *bitmap, uint32_t id);
-bool bondp_bitmap_is_set(bondp_bitmap_t *bitmap, uint32_t id);
-#endif // BONDP_BITMAP_H
\ No newline at end of file
diff --git a/src/urma/lib/urma/bond/bondp_provider_ops.c b/src/urma/lib/urma/bond/bondp_provider_ops.c
index 22a2c0e8b4f1a098da6e212c0cc03bb53c3b776d..e6fc1c5a2291ce5a5b975316709335249b2f335e 100644
--- a/src/urma/lib/urma/bond/bondp_provider_ops.c
+++ b/src/urma/lib/urma/bond/bondp_provider_ops.c
@@ -291,13 +291,9 @@ static void bondp_uninit_v_ctx(bondp_context_t *bond_ctx)
 
 static int bondp_init_ctx_table(bondp_context_t *bond_ctx)
 {
-    if (bondp_bitmap_init(&bond_ctx->token_id_bitmap, BONDP_MAX_NUM_SEGS + 1)) {
-        URMA_LOG_ERR("Failed to create token_id_bitmap\n");
-        return -1;
-    }
     if (bdp_p_vjetty_id_table_create(&bond_ctx->p_vjetty_id_table, BONDP_MAX_NUM_JETTYS)) {
         URMA_LOG_ERR("Failed to create p_vjetty_id_table\n");
-        goto FREE_TOKEN_ID_BITMAP;
+        return -1;
     }
     if (bdp_r_p2v_jetty_id_table_create(&bond_ctx->remote_p2v_jetty_id_table, BONDP_MAX_NUM_JETTYS)) {
         URMA_LOG_ERR("Failed to create remote_p2v_jetty_id_table\n");
@@ -313,8 +309,6 @@ FREE_V_PTOKEN_ID_TABLE:
     (void)bdp_r_p2v_jetty_id_table_destroy(&bond_ctx->remote_p2v_jetty_id_table);
 FREE_P_VJETTY_ID_TABLE:
     bdp_p_vjetty_id_table_destroy(&bond_ctx->p_vjetty_id_table);
-FREE_TOKEN_ID_BITMAP:
-    bondp_bitmap_uninit(&bond_ctx->token_id_bitmap);
     return -1;
 }
 
@@ -323,7 +317,6 @@ static void bondp_uninit_ctx_table(bondp_context_t *bond_ctx)
     bdp_r_v2p_token_id_table_destroy(&bond_ctx->remote_v2p_token_id_table);
     bdp_r_p2v_jetty_id_table_destroy(&bond_ctx->remote_p2v_jetty_id_table);
     bdp_p_vjetty_id_table_destroy(&bond_ctx->p_vjetty_id_table);
-    bondp_bitmap_uninit(&bond_ctx->token_id_bitmap);
 }
 
 static bondp_context_t* bondp_create_ctx(void)
diff --git a/src/urma/lib/urma/bond/bondp_segment.c b/src/urma/lib/urma/bond/bondp_segment.c
index c32efe88118c15bad44c0d85b650c787c82bd464..fe914247877390b8c90fe6f2ddc3c8a8a99958ee 100644
--- a/src/urma/lib/urma/bond/bondp_segment.c
+++ b/src/urma/lib/urma/bond/bondp_segment.c
@@ -105,41 +105,23 @@ urma_target_seg_t *bondp_find_vtseg_by_va(uint64_t va)
     return va_vtseg_info->vtseg;
 }
 
+/*
+ * Since the actual token ID for bonding device is allocated by the kernel mode
+ * during segment registration, the function only returns an empty token ID to
+ * make api useable
+ */
 urma_token_id_t *bondp_alloc_token_id(urma_context_t *ctx)
 {
-    urma_token_id_t *token = NULL;
-    bondp_context_t *bdp_ctx = CONTAINER_OF_FIELD(ctx, bondp_context_t, v_ctx);
-
-    if (!is_valid_ctx(bdp_ctx)) {
-        URMA_LOG_ERR("Invalid bdp_ctx");
-        return NULL;
-    }
-
-    token = calloc(1, sizeof(urma_token_id_t));
+    urma_token_id_t *token = calloc(1, sizeof(urma_token_id_t));
     if (token == NULL) {
         return NULL;
     }
-    if (bondp_bitmap_alloc_idx(&bdp_ctx->token_id_bitmap, &token->token_id)) {
-        URMA_LOG_ERR("Failed to alloc token id.\n");
-        free(token);
-        return NULL;
-    }
     token->urma_ctx = ctx;
     return token;
 }
 
 urma_status_t bondp_free_token_id(urma_token_id_t *token_id)
 {
-    bondp_context_t *bdp_ctx = CONTAINER_OF_FIELD(token_id->urma_ctx, bondp_context_t, v_ctx);
-
-    if (!is_valid_ctx(bdp_ctx)) {
-        URMA_LOG_ERR("Invalid bdp_ctx");
-        return URMA_FAIL;
-    }
-    if (bondp_bitmap_free_idx(&bdp_ctx->token_id_bitmap, token_id->token_id)) {
-        URMA_LOG_ERR("Failed to free idx, tid: %u.\n", token_id->token_id);
-        return URMA_FAIL;
-    }
     free(token_id);
     return URMA_SUCCESS;
 }
diff --git a/src/urma/lib/urma/bond/bondp_types.h b/src/urma/lib/urma/bond/bondp_types.h
index f26baa247c8bb75b25fa7fc456f5e7cfe16d4ef8..393b1a4f48944f507b44e0c92af05668d75b4dfb 100644
--- a/src/urma/lib/urma/bond/bondp_types.h
+++ b/src/urma/lib/urma/bond/bondp_types.h
@@ -14,7 +14,6 @@
 #include "urma_types.h"
 #include "urma_ubagg.h"
 #include "bondp_hash_table.h"
-#include "bondp_bitmap.h"
 #include "topo_info.h"
 
 #define BONDP_MAX_NUM_JFRS (10240)
@@ -79,7 +78,6 @@ typedef struct bondp_context {
     /* and valid devices are confirmed through the non-null p_ctxs[i]. */
     int dev_num;
     topo_map_t *topo_map;               /* Only valid when the node is in matrix server */
-    bondp_bitmap_t token_id_bitmap;     /* alloc token_id used by urma_jetty */
     bondp_hash_table_t tjetty_id_table; /* match target jetty's pjetty_id to vjetty_id */
     /* Record the mapping from the locally created jetty's pjetty.jetty_id.id to the vjetty.jetty_id.id, */
     /* used to restore the local_id in CR. */
diff --git a/test/urma/bond/bondp_bitmap_gtest.cpp b/test/urma/bond/bondp_bitmap_gtest.cpp
deleted file mode 100644
index cf2707bab10b84b6564428e01fd681025effcd3b..0000000000000000000000000000000000000000
--- a/test/urma/bond/bondp_bitmap_gtest.cpp
+++ /dev/null
@@ -1,45 +0,0 @@
-/*
- * SPDX-License-Identifier: MIT
- * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
- * Description: bitmap gtest source file
- * Author: Chen Wen
- * Create: 2025-07-10
- * Note:
- * History:
- */
-
-#include "bondp_gtest_basic.h"
-
-extern "C" {
-#include "bondp_bitmap.h"
-}
-
-class BondpBitmapGTest : public BondpBasicGTest {
-public:
-protected:
-};
-
-TEST_F(BondpBitmapGTest, Bondp_BitmapAllocTest){
-    uint32_t bitmap_size = 32;
-    bondp_bitmap_t *bitmap = bondp_bitmap_alloc(bitmap_size);
-    EXPECT_NE((void *)NULL, (void *)bitmap);
-}
-
-TEST_F(BondpBitmapGTest, Bondp_BitmapUsedIdTest){
-    uint32_t bitmap_size = 32;
-    bondp_bitmap_t *bitmap = bondp_bitmap_alloc(bitmap_size);
-    uint32_t id=16;
-    int ret;
-    ret = bondp_bitmap_use_id(bitmap,id);
-    EXPECT_EQ(0,ret);
-    ret = bondp_bitmap_use_id(bitmap,id);
-    EXPECT_EQ(-1,ret);
-    ret = bondp_bitmap_free_idx(bitmap,id);
-    EXPECT_EQ(0,ret);
-    ret = bondp_bitmap_free_idx(bitmap,id);
-    EXPECT_NE(0,ret);
-    bondp_bitmap_free(bitmap);
-}
-
-
-
