diff --git a/src/urma/tools/urma_perftest/perftest_parameters.h b/src/urma/tools/urma_perftest/perftest_parameters.h
index e174fc55604c85fe51fb0c57f4a4ce440a6d16f4..3672186faf4dd152ef035ba08c921b8a3fc66b65 100644
--- a/src/urma/tools/urma_perftest/perftest_parameters.h
+++ b/src/urma/tools/urma_perftest/perftest_parameters.h
@@ -38,7 +38,7 @@
 #define PERFTEST_DEF_NUM_JETTYS   (1)
 #define PERFTEST_DEF_ATOMIC_SIZE   (8)
 #define PERFTEST_SIZE_CQ_MOD_LIMIT (8192)
-#define PERFTEST_BW_NO_PEAK_INTERS (20000)
+#define PERFTEST_BW_NO_PEAK_INTERS (50000)
 
 #define PERFTEST_DEF_CACHE_LINE_SIZE (64)
 #define PERFTEST_PAGE_SIZE (4096)
