diff --git a/src/urma/tools/urma_perftest/perftest_run_test.c b/src/urma/tools/urma_perftest/perftest_run_test.c
index 7338e8834f1ef5f7c0de94ac4b4101dd975988f4..4f1495a2de1d9f1c834fa577931fb6b0706b942d 100644
--- a/src/urma/tools/urma_perftest/perftest_run_test.c
+++ b/src/urma/tools/urma_perftest/perftest_run_test.c
@@ -1082,6 +1082,9 @@ static int prepare_jfr_wr(perftest_context_t *ctx, perftest_config_t *cfg)
     uint32_t size_per_jetty = cfg->jfr_depth / cfg->jfr_post_list;
     uint32_t local_sge_idx;
 
+    if (cfg->share_jfr == true) {
+        size_per_jetty /= cfg->jettys;
+    }
     run_ctx->rposted = (int)(size_per_jetty * cfg->jfr_post_list);
 
     if (alloc_jfr_ctx_buffer(ctx, cfg) != 0) {
