diff --git a/src/urma/common/include/ub_hmap.h b/src/urma/common/include/ub_hmap.h
index 1895fb1b316a150b9b1987f3ab78bced5504f52f..9827e27ca51b5e47f97a4072db5ab0ba9e57063e 100644
--- a/src/urma/common/include/ub_hmap.h
+++ b/src/urma/common/include/ub_hmap.h
@@ -117,16 +117,16 @@ static inline uint32_t ub_hmap_count(const struct ub_hmap *hmap)
 
 #define HMAP_FOR_EACH_WITH_HASH(NODE, MEMBER, HASH, TABLE)                       \
     for (INIT_CONTAINER_PTR(NODE, ub_hmap_first_with_hash(TABLE, HASH), MEMBER); \
-        ((NODE != OBJ_CONTAINING(NULL, NODE, MEMBER)) || (NODE = NULL));           \
+        (((NODE) != OBJ_CONTAINING(NULL, NODE, MEMBER)) || ((NODE) = NULL));           \
         ASSIGN_CONTAINER_PTR(NODE, ub_hmap_next_with_hash(&(NODE)->MEMBER, HASH), MEMBER))
 
 #define HMAP_FOR_EACH(NODE, MEMBER, TABLE) for (INIT_CONTAINER_PTR(NODE, ub_hmap_first(TABLE), MEMBER);       \
-        ((NODE != OBJ_CONTAINING(NULL, NODE, MEMBER)) || (NODE = NULL)); \
+        (((NODE) != OBJ_CONTAINING(NULL, NODE, MEMBER)) || ((NODE) = NULL)); \
         ASSIGN_CONTAINER_PTR(NODE, ub_hmap_next(TABLE, &(NODE)->MEMBER), MEMBER))
 
 #define HMAP_FOR_EACH_SAFE(NODE, NEXT, MEMBER, HMAP)                           \
     for (INIT_CONTAINER_PTR(NODE, ub_hmap_first(HMAP), MEMBER);                \
-        (((NODE != OBJ_CONTAINING(NULL, NODE, MEMBER)) || (NODE = NULL)) ?       \
+        ((((NODE) != OBJ_CONTAINING(NULL, NODE, MEMBER)) || ((NODE) = NULL)) ?       \
         INIT_CONTAINER_PTR(NEXT, ub_hmap_next(HMAP, &(NODE)->MEMBER), MEMBER), 1 : 0); \
         (NODE) = (NEXT))
 
diff --git a/src/urma/common/include/ub_list.h b/src/urma/common/include/ub_list.h
index 2d063a931fe6f0d3fe720ef4647f21e45b58c8d6..92bc4dd339459bc376892178c696cd253dbcc0e5 100644
--- a/src/urma/common/include/ub_list.h
+++ b/src/urma/common/include/ub_list.h
@@ -187,11 +187,11 @@ static inline int ub_list_insert_ordered_without_repetition(struct ub_list *list
 
 /* for list that each node has a different data struct */
 #define UB_LIST_FOR_EACH_BY_NODE(NODE, HEAD) \
-    for ((NODE = HEAD->next); ((NODE != HEAD) && (NODE != NULL)); (NODE = NODE->next))
+    for (((NODE) = (HEAD)->next); (((NODE) != (HEAD)) && ((NODE) != NULL)); ((NODE) = (NODE)->next))
 
 #define UB_LIST_FOR_EACH_BY_NODE_SAFE(NODE, NEXT_NODE, HEAD) \
-    for ((NODE = HEAD->next); ((NODE != HEAD) && (NODE != NULL) && (NEXT_NODE = NODE->next, 1)); \
-        (NODE = NEXT_NODE))
+    for (((NODE) = (HEAD)->next); (((NODE) != (HEAD)) && ((NODE) != NULL) && ((NEXT_NODE) = (NODE)->next, 1)); \
+        ((NODE) = (NEXT_NODE)))
 
 /* used to get first node of a list, if list is empty, ITER will be set to NULL */
 #define UB_LIST_FIRST_NODE(ITER, MEMBER, LIST)                              \
