diff --git a/src/urma/lib/uvs/core/include/uvs_api.h b/src/urma/lib/uvs/core/include/uvs_api.h
index 0217de59ee3af8098ed7d7645efe23d30bbc6226..447de2fbb3049a0aadf96862c761b28087e9fe80 100644
--- a/src/urma/lib/uvs/core/include/uvs_api.h
+++ b/src/urma/lib/uvs/core/include/uvs_api.h
@@ -47,9 +47,10 @@ typedef struct uvs_route_list {
 /**
  * Create an aggregation device in UVS.
  * @param[in] agg_eid  EID of the aggregation device to be created.
+ * @param[in] dev_name  Name of the aggregation device to be created.
  * @return 0 on success, other value on error.
  */
-int uvs_create_agg_dev(uvs_eid_t *agg_eid);
+int uvs_create_agg_dev(uvs_eid_t *agg_eid, const char *dev_name);
 
 /**
  * Delete an aggregation device from UVS.
@@ -96,4 +97,4 @@ int uvs_get_route_list(const uvs_route_t *route, uvs_route_list_t *route_list);
 }
 #endif
 
-#endif
\ No newline at end of file
+#endif
diff --git a/src/urma/lib/uvs/core/tpsa_api.c b/src/urma/lib/uvs/core/tpsa_api.c
index ea3ff436fe9fd9f686b84cb76a5c77a525d44713..e5fa9b23414d0ac39edb1473e9341259c3da83ae 100644
--- a/src/urma/lib/uvs/core/tpsa_api.c
+++ b/src/urma/lib/uvs/core/tpsa_api.c
@@ -18,16 +18,23 @@
 
 #define UVS_MAX_TOPO_NUM 16
 
-int uvs_create_agg_dev(uvs_eid_t *agg_eid)
+int uvs_create_agg_dev(uvs_eid_t *agg_eid, const char *dev_name)
 {
     int ret = 0;
+    size_t dev_name_len;
 
-    if (agg_eid == NULL) {
+    if (agg_eid == NULL || dev_name == NULL) {
         TPSA_LOG_ERR("Invalid parameter.\n");
         return -EINVAL;
     }
 
-    ret = uvs_ubagg_ioctl_create_agg_dev(agg_eid);
+    dev_name_len = strnlen(dev_name, UVS_MAX_DEV_NAME_LEN);
+    if (dev_name_len == 0 || dev_name_len >= UVS_MAX_DEV_NAME_LEN) {
+        TPSA_LOG_ERR("Invalid parameter.\n");
+        return -EINVAL;
+    }
+
+    ret = uvs_ubagg_ioctl_create_agg_dev(agg_eid, dev_name);
     if (ret != 0) {
         TPSA_LOG_ERR("failed to create agg dev in ubagg, ret = %d.\n", ret);
         return ret;
diff --git a/src/urma/lib/uvs/core/uvs_ubagg_ioctl.c b/src/urma/lib/uvs/core/uvs_ubagg_ioctl.c
index 8851f834f51dfef22a3f8ee4ff288af7b2edd1df..9d69d51c861455cd0aa35ca39312d65a5d239239 100644
--- a/src/urma/lib/uvs/core/uvs_ubagg_ioctl.c
+++ b/src/urma/lib/uvs/core/uvs_ubagg_ioctl.c
@@ -20,13 +20,26 @@
 #define UVS_UBAGG_DEVICE_PATH "/dev/ubagg"
 #define UVS_UBCORE_DEVICE_PATH "/dev/ubcore/ubcore"
 
-int uvs_ubagg_ioctl_create_agg_dev(uvs_eid_t *agg_eid)
+int uvs_ubagg_ioctl_create_agg_dev(uvs_eid_t *agg_eid, const char *dev_name)
 {
     struct uvs_ubagg_create_dev_arg args = {0};
     struct uvs_ubagg_cmd_hdr hdr = {0};
     int ret;
+    size_t dev_name_len;
+
+    if (agg_eid == NULL || dev_name == NULL) {
+        TPSA_LOG_ERR("Invalid parameter.\n");
+        return -1;
+    }
+
+    dev_name_len = strnlen(dev_name, UVS_MAX_DEV_NAME_LEN);
+    if (dev_name_len == 0 || dev_name_len >= UVS_MAX_DEV_NAME_LEN) {
+        TPSA_LOG_ERR("Invalid parameter.\n");
+        return -1;
+    }
 
     args.in.agg_eid = *agg_eid;
+    (void)snprintf(args.in.dev_name, UVS_MAX_DEV_NAME_LEN, "%s", dev_name);
 
     hdr.command = UVS_UBAGG_CMD_CREATE_DEV;
     hdr.args_addr = (uint64_t)(uintptr_t)&args;
diff --git a/src/urma/lib/uvs/core/uvs_ubagg_ioctl.h b/src/urma/lib/uvs/core/uvs_ubagg_ioctl.h
index 59080e59653ab8d037c8a188f5fa45de049f9208..ad0f108042cf3959673a021909c7552a1223d7a0 100644
--- a/src/urma/lib/uvs/core/uvs_ubagg_ioctl.h
+++ b/src/urma/lib/uvs/core/uvs_ubagg_ioctl.h
@@ -37,6 +37,7 @@ struct uvs_ubagg_cmd_hdr {
 struct uvs_ubagg_create_dev_arg {
     struct {
         uvs_eid_t agg_eid;
+        char dev_name[UVS_MAX_DEV_NAME_LEN];
     } in;
 };
 
@@ -62,7 +63,7 @@ struct uvs_ubagg_set_topo_info {
     } in;
 };
 
-int uvs_ubagg_ioctl_create_agg_dev(uvs_eid_t *agg_eid);
+int uvs_ubagg_ioctl_create_agg_dev(uvs_eid_t *agg_eid, const char *dev_name);
 int uvs_ubagg_ioctl_delete_agg_dev(uvs_eid_t *agg_eid);
 int uvs_ubagg_ioctl_get_dev_name_by_eid(uvs_eid_t *eid, char *buf, size_t len);
 int uvs_ubagg_ioctl_set_topo(void *topo_info, int topo_num);
@@ -70,4 +71,4 @@ int uvs_ubcore_ioctl_get_topo(void *topo_map);
 int uvs_ubcore_ioctl_set_topo(void *topo_info, int topo_num);
 int uvs_ubcore_ioctl_get_route_list(const uvs_route_t *route, uvs_route_list_t *route_list);
 
-#endif // UVS_UBAGG_IOCTL_H
\ No newline at end of file
+#endif // UVS_UBAGG_IOCTL_H
