diff --git a/src/urma/lib/urma/bond/bondp_datapath.c b/src/urma/lib/urma/bond/bondp_datapath.c
index b4070375fc2a4303cf906a6842552b8bf15e2c2f..a58fb2b57da7ce144750136eb509bdb97e72a0e8 100644
--- a/src/urma/lib/urma/bond/bondp_datapath.c
+++ b/src/urma/lib/urma/bond/bondp_datapath.c
@@ -747,7 +747,7 @@ static cr_convert_ret_t handle_fake_cr_with_store(bondp_context_t *bdp_ctx, int
             break;
         }
     }
-    if (all_reported) {
+    if (all_reported && !comp->modify_to_error) {
         cr->local_id = get_comp_urma_jetty_id(comp)->id;
         /* Caller should copy this CR to output array. */
         put_comp(comp);
diff --git a/src/urma/lib/urma/bond/bondp_health_check.c b/src/urma/lib/urma/bond/bondp_health_check.c
index 1d29650249d529c7c0f42b35dd1bcbb275abd5b7..a5cefdc27ff3b97cbd4f73ea9651c04715af4561 100644
--- a/src/urma/lib/urma/bond/bondp_health_check.c
+++ b/src/urma/lib/urma/bond/bondp_health_check.c
@@ -1252,8 +1252,8 @@ void bondp_health_update_active_idx(bondp_context_t *bdp_ctx, bondp_target_jetty
         }
         int old = task->active_local_idx;
         task->active_local_idx = new_active_idx;
+        task->bondp_jetty->pjettys_error_done[new_active_idx] &= (uint8_t)(~PJETTY_FLUSH_ERROR_DONE);
         if (new_active_idx == task->primary_local_idx) {
-            task->bondp_jetty->pjettys_error_done[new_active_idx] &= (uint8_t)(~PJETTY_FLUSH_ERROR_DONE);
             task->fallback_task.relink_done = false;
             task->fallback_task.pending = false;
             task->fallback_task.local_rebuilt = false;
