diff --git a/src/urma/lib/urma/core/urma_device.c b/src/urma/lib/urma/core/urma_device.c
index a5be59b75a39f208d04929f3db8718c014fd9c92..70dfdf0fe8b96d266f5cc8583f7eee111ea4c549 100644
--- a/src/urma/lib/urma/core/urma_device.c
+++ b/src/urma/lib/urma/core/urma_device.c
@@ -341,8 +341,11 @@ static void urma_read_sysfs_dev_attrs(urma_sysfs_dev_t *sysfs_dev)
     urma_parse_string(sysfs_dev->sysfs_path, "driver_name", sysfs_dev->driver_name, URMA_MAX_NAME);
 
     sysfs_dev->transport_type = (urma_transport_type_t)urma_parse_value_u32(sysfs_dev->sysfs_path, "transport_type");
-    sysfs_dev->vendor_id = urma_parse_value_u16(sysfs_dev->sysfs_path, "device/vendor");
-    sysfs_dev->device_id = urma_parse_value_u16(sysfs_dev->sysfs_path, "device/device");
+    /* bongding device does not have vendor and device char device file */
+    sysfs_dev->vendor_id = (urma_is_bonding_dev(sysfs_dev->dev_name)) ? 0 :
+        urma_parse_value_u16(sysfs_dev->sysfs_path, "device/vendor");
+    sysfs_dev->device_id = (urma_is_bonding_dev(sysfs_dev->dev_name)) ? 0 :
+        urma_parse_value_u16(sysfs_dev->sysfs_path, "device/device");
 
     if (urma_query_device_attr(sysfs_dev) == 0) {
         return;
diff --git a/src/urma/lib/urma/core/urma_private.h b/src/urma/lib/urma/core/urma_private.h
index be7aff8cd45e404df3ad3c821b5d1a50cb75df2b..1ee9e874f594fea23c50771221abbed90464b2c1 100644
--- a/src/urma/lib/urma/core/urma_private.h
+++ b/src/urma/lib/urma/core/urma_private.h
@@ -17,6 +17,7 @@
 #include "urma_provider.h"
 
 #define URMA_MAX_SYSFS_PATH 256
+#define URMA_UBAGG_DEV_PREFIX "bonding_dev"
 
 typedef struct urma_driver {
     struct urma_provider_ops *ops;
@@ -49,4 +50,14 @@ void urma_uninit_jetty_cfg(urma_jetty_cfg_t *p);
 int urma_query_eid(urma_device_t *dev, uint32_t eid_index, urma_eid_t *eid);
 int urma_open_cdev(char *path);
 
+static inline bool urma_is_bonding_dev(char *dev_name)
+{
+    if (strnlen(dev_name, URMA_MAX_NAME) >= URMA_MAX_NAME) {
+        return false;
+    }
+    
+    return memcmp(dev_name, URMA_UBAGG_DEV_PREFIX,
+        strlen(URMA_UBAGG_DEV_PREFIX)) == 0;
+}
+
 #endif // URMA_PRIVATE_H
