diff --git a/src/urma/lib/urma/bond/bondp_api.c b/src/urma/lib/urma/bond/bondp_api.c
index c36a5ba36a355b63b6e2d421b5c8ecd375452136..f1eeacb39e6c4b68db330fc4284773d27a5cab44 100644
--- a/src/urma/lib/urma/bond/bondp_api.c
+++ b/src/urma/lib/urma/bond/bondp_api.c
@@ -257,7 +257,7 @@ DEL_P_VJFS_ID:
         if (bdp_jfs->p_jfs[j] == NULL) {
             continue;
         }
-        urma_jfr_id_t pjfs_id = bdp_jfs->p_jfs[i]->jfs_id;
+        urma_jfr_id_t pjfs_id = bdp_jfs->p_jfs[j]->jfs_id;
         (void)bdp_p_vjetty_id_table_del_without_lock(&bdp_ctx->p_vjetty_id_table, pjfs_id, JFS);
     }
     pthread_rwlock_unlock(&bdp_ctx->p_vjetty_id_table.lock);
