diff --git a/src/urma/lib/urma/bond/bondp_connection.c b/src/urma/lib/urma/bond/bondp_connection.c
index 0cc9e6878d49c8959372a21d1ffdacaa7aa1ced4..00ca0269bfee24b7ea59adae9a34b42d0803ae8c 100644
--- a/src/urma/lib/urma/bond/bondp_connection.c
+++ b/src/urma/lib/urma/bond/bondp_connection.c
@@ -7,28 +7,17 @@
  * Note:
  * History: 2025-03-06  Create file
  */
-#include "bondp_connection.h"
+#include <stdlib.h>
+
 #include "bondp_types.h"
 #include "ub_hash.h"
 #include "ub_util.h"
 #include "urma_log.h"
 
-#define BDP_V_CONN_HASH_BASIS (0x87820129)
-
-void init_v_conn_on_send(bdp_v_conn_t *v_conn, void *target_vjetty, int target_dev_num)
-{
-    if (v_conn == NULL || target_vjetty == NULL) {
-        URMA_LOG_ERR("Invalid param\n");
-        return;
-    }
-    bondp_target_jetty_t *bdp_tjetty = target_vjetty;
-    v_conn->target_vjetty = bdp_tjetty;
-}
+#include "bondp_connection.h"
 
 int bdp_v_conn_init(bdp_v_conn_t *v_conn)
 {
-    v_conn->target_vjetty = NULL;
-    v_conn->msn = 0;
     if (bdp_slide_wnd_init(&v_conn->recv_wnd, BONDP_MAX_BITMAP_SIZE, BONDP_RECV_WND_SIZE, 0)) {
         URMA_LOG_ERR("Failed to init slide window in bdp_v_conn_table_add\n");
         return -1;
@@ -73,36 +62,6 @@ int bdp_v_conn_table_create(bondp_hash_table_t *tbl, uint32_t size)
     return bondp_hash_table_create(tbl, size, v_conn_comp, v_conn_free, v_conn_hash);
 }
 
-int bdp_v_conn_table_add_on_send(bondp_hash_table_t *tbl, urma_jetty_id_t *target_id,
-                                 void *target_vjetty, int target_dev_num, bdp_v_conn_t **v_conn_out, bool is_standalone)
-{
-    hmap_node_t *node = NULL;
-    uint32_t hash = tbl->hash_f(target_id);
-    bdp_v_conn_node_t *v_conn_node = NULL;
-
-    node = bondp_hash_table_lookup_without_lock(tbl, target_id, hash);
-    if (node) {
-        return BONDP_HASH_MAP_COLLIDE_ERROR;
-    }
-    v_conn_node = malloc(sizeof(bdp_v_conn_node_t));
-    if (v_conn_node == NULL) {
-        return BONDP_HASH_MAP_ALLOC_ERROR;
-    }
-    if (!is_standalone) {
-        if (bdp_v_conn_init(&v_conn_node->v_conn)) {
-            goto FREE_VCONN_NODE;
-        }
-    }
-    init_v_conn_on_send(&v_conn_node->v_conn, target_vjetty, target_dev_num);
-    v_conn_node->target_id = *target_id;
-    bondp_hash_table_add_with_hash(tbl, &v_conn_node->hmap_node, hash);
-    *v_conn_out = &v_conn_node->v_conn;
-    return 0;
-FREE_VCONN_NODE:
-    free(v_conn_node);
-    return BONDP_HASH_MAP_ALLOC_ERROR;
-}
-
 int bdp_v_conn_table_add_on_recv(bondp_hash_table_t *tbl, urma_jetty_id_t *target_id, bdp_v_conn_t **v_conn_out,
                                  bool is_standalone)
 {
@@ -114,7 +73,7 @@ int bdp_v_conn_table_add_on_recv(bondp_hash_table_t *tbl, urma_jetty_id_t *targe
     if (node) {
         return BONDP_HASH_MAP_COLLIDE_ERROR;
     }
-    v_conn_node = malloc(sizeof(bdp_v_conn_node_t));
+    v_conn_node = calloc(1, sizeof(bdp_v_conn_node_t));
     if (v_conn_node == NULL) {
         return BONDP_HASH_MAP_ALLOC_ERROR;
     }
diff --git a/src/urma/lib/urma/bond/bondp_connection.h b/src/urma/lib/urma/bond/bondp_connection.h
index 57b62419a68413eb9bd23ef52c31928c1ce897dd..1b28311db6759dc115f95b260c1a8dfb03cf3428 100644
--- a/src/urma/lib/urma/bond/bondp_connection.h
+++ b/src/urma/lib/urma/bond/bondp_connection.h
@@ -21,16 +21,11 @@
  * Virtual connection is used to implement de-duplication between two vjettys.
  */
 typedef struct bondp_v_connection {
-    /* only valid on sender side */
-    bondp_target_jetty_t *target_vjetty;
     /* de-duplication */
-    uint32_t msn;
     bdp_slide_wnd_t recv_wnd;
     bool is_init;
 } bdp_v_conn_t;
 
-void init_v_conn_on_send(bdp_v_conn_t *v_conn, void *target_vjetty, int target_dev_num);
-
 typedef struct bdp_v_conn_node {
     hmap_node_t hmap_node;
     urma_jetty_id_t target_id; /* key */
@@ -41,9 +36,6 @@ int bdp_v_conn_table_create(bondp_hash_table_t *tbl, uint32_t size);
 
 bdp_v_conn_t *bdp_v_conn_table_lookup(bondp_hash_table_t *tbl, urma_jetty_id_t *target_id);
 
-int bdp_v_conn_table_add_on_send(bondp_hash_table_t *tbl, urma_jetty_id_t *target_id,
-                                 void *target_vjetty, int target_dev_num, bdp_v_conn_t **v_conn_out, bool is_standalone);
-
 int bdp_v_conn_table_add_on_recv(bondp_hash_table_t *tbl, urma_jetty_id_t *target_id,
                                  bdp_v_conn_t **v_conn_out, bool is_standalone);
 
diff --git a/src/urma/lib/urma/bond/bondp_datapath.c b/src/urma/lib/urma/bond/bondp_datapath.c
index 0c6c1ac74fe02d12a5a1c75372c59aa6ada4efed..855a637315cfda370cc0bfbcdcada60ada232137 100644
--- a/src/urma/lib/urma/bond/bondp_datapath.c
+++ b/src/urma/lib/urma/bond/bondp_datapath.c
@@ -192,33 +192,6 @@ static urma_status_t post_send_check_wr_list_valid(bondp_comp_t *bdp_send_comp,
     return URMA_SUCCESS;
 }
 
-static bdp_v_conn_t *get_v_conn_on_send(bondp_comp_t *bdp_comp, bondp_target_jetty_t *bdp_tjetty)
-{
-    urma_jetty_id_t *comp_jetty_id = get_comp_urma_jetty_id(bdp_comp);
-    if (comp_jetty_id == NULL) {
-        return NULL;
-    }
-    urma_jetty_id_t *vtjetty_id = &bdp_tjetty->v_tjetty.id;
-    bdp_v_conn_t *v_conn = bdp_v_conn_table_lookup(&bdp_comp->v_conn_table, vtjetty_id);
-    if (!v_conn) {
-        int ret = bdp_v_conn_table_add_on_send(&bdp_comp->v_conn_table, vtjetty_id,
-                                               bdp_tjetty, bdp_tjetty->target_dev_num, &v_conn,
-                                               is_single_dev_mode(bdp_comp->bondp_ctx));
-        if (ret != 0) {
-            URMA_LOG_ERR("Failed to create v_conn for vjetty, ret: %d, "
-                         "[" URMA_JETTY_ID_FMT " -> " URMA_JETTY_ID_FMT "]\n",
-                         ret, URMA_JETTY_ID_ARGS(comp_jetty_id), URMA_JETTY_ID_ARGS(vtjetty_id));
-            return NULL;
-        }
-    }
-    /* v_conn is not null */
-    /* If this v_conn is created by handle_recv, then we need to initialize it */
-    if (v_conn->target_vjetty == NULL) {
-        init_v_conn_on_send(v_conn, bdp_tjetty, bdp_tjetty->target_dev_num);
-    }
-    return v_conn;
-}
-
 static urma_status_t bondp_post_send_wr_no_store(bondp_comp_t *bdp_comp,
                                                  const urma_jfs_wr_t *wr, urma_jfs_wr_t **bad_wr)
 {
@@ -235,11 +208,6 @@ static urma_status_t bondp_post_send_wr_no_store(bondp_comp_t *bdp_comp,
         return URMA_EINVAL;
     }
 
-    bdp_v_conn_t *v_conn = get_v_conn_on_send(bdp_comp, bdp_tjetty);
-    if (v_conn == NULL) {
-        return URMA_FAIL;
-    }
-
     int send_idx = -1, target_idx = -1;
     if (!wr->flag.bs.has_drv_ext) {
         ret = schedule_send(wr->tjetty, bdp_comp, &send_idx, &target_idx, NULL);
@@ -260,7 +228,7 @@ static urma_status_t bondp_post_send_wr_no_store(bondp_comp_t *bdp_comp,
         if (ret != 0) {
             return ret;
         }
-        ret = convert_jfs_vwr_to_pwr(pwr, send_idx, target_idx, bdp_comp, v_conn);
+        ret = convert_jfs_vwr_to_pwr(pwr, send_idx, target_idx, bdp_comp);
         if (ret != 0) {
             return ret;
         }
@@ -291,11 +259,6 @@ static urma_status_t bondp_post_send_wr_and_store(bondp_comp_t *bdp_comp, urma_j
         return URMA_EINVAL;
     }
 
-    bdp_v_conn_t *v_conn = get_v_conn_on_send(bdp_comp, bdp_tjetty);
-    if (v_conn == NULL) {
-        return URMA_FAIL;
-    }
-
     int send_idx = 0, target_idx = 0;
     if (!wr->flag.bs.has_drv_ext) {
         ret = schedule_send(wr->tjetty, bdp_comp, &send_idx, &target_idx, NULL);
@@ -317,7 +280,7 @@ static urma_status_t bondp_post_send_wr_and_store(bondp_comp_t *bdp_comp, urma_j
     }
     wr_entry->user_ctx = wr->user_ctx;
     wr_entry->bdp_comp = bdp_comp;
-    wr_entry->v_conn = v_conn;
+    wr_entry->target_vjetty = bdp_tjetty;
     wr_entry->send_idx = send_idx;
     wr_entry->target_idx = target_idx;
 
@@ -329,7 +292,7 @@ static urma_status_t bondp_post_send_wr_and_store(bondp_comp_t *bdp_comp, urma_j
     }
 
     add_vwr_use_cnt(pwr);
-    ret = convert_jfs_vwr_to_pwr(pwr, send_idx, target_idx, bdp_comp, v_conn);
+    ret = convert_jfs_vwr_to_pwr(pwr, send_idx, target_idx, bdp_comp);
     if (ret != 0) {
         URMA_LOG_ERR("Failed to convert jfs wr\n");
         goto RELEASE_WR;
@@ -344,7 +307,7 @@ static urma_status_t bondp_post_send_wr_and_store(bondp_comp_t *bdp_comp, urma_j
     return URMA_SUCCESS;
 
 RELEASE_WR:
-    convert_jfs_pwr_to_vwr_resend(pwr, &wr_entry->v_conn->target_vjetty->v_tjetty);
+    convert_jfs_pwr_to_vwr_resend(pwr, &wr_entry->target_vjetty->v_tjetty);
     release_vwr_use_cnt(pwr);
 FREE_PWR:
     free_jfs_wr(pwr);
@@ -660,7 +623,7 @@ static int resend_jfs_wr(bondp_comp_t *bdp_comp, jfs_wr_entry_t *wr_entry, int s
     wr_entry->send_idx = send_idx;
     wr_entry->target_idx = target_idx;
     urma_jfs_wr_t *wr = &wr_entry->wr;
-    urma_target_jetty_t *vtjetty = &wr_entry->v_conn->target_vjetty->v_tjetty;
+    urma_target_jetty_t *vtjetty = &wr_entry->target_vjetty->v_tjetty;
     convert_jfs_pwr_to_vwr_resend(wr, vtjetty);
     convert_jfs_vwr_to_pwr_for_resend(wr, send_idx, target_idx);
 
@@ -791,12 +754,12 @@ static cr_convert_ret_t handle_send_cr_with_store(bondp_context_t *bdp_ctx, bond
         return CONVERT_SKIP;
     }
 
-    if (false && is_failover_cr(cr) && !bdp_comp->modify_to_error) {
+    if (is_failover_cr(cr) && !bdp_comp->modify_to_error) {
         (void)pthread_spin_lock(&bdp_comp->send_lock);
         bdp_comp->valid[send_idx] = false;
 
         int new_send_idx = -1, new_target_idx = -1;
-        if (schedule_send(&wr_entry->v_conn->target_vjetty->v_tjetty, bdp_comp,
+        if (schedule_send(&wr_entry->target_vjetty->v_tjetty, bdp_comp,
                           &new_send_idx, &new_target_idx, NULL) != 0) {
             /*
              * When all ports are invalid and no port is available to resend the wr,
@@ -824,10 +787,10 @@ static cr_convert_ret_t handle_send_cr_with_store(bondp_context_t *bdp_ctx, bond
                 URMA_LOG_ERR("Failed to resend jfs wr, wr_id: %lu\n", wr_id);
             }
         }
-        bondp_health_notify_datapath_link_fail(bdp_comp->bondp_ctx, wr_entry->v_conn->target_vjetty,
+        bondp_health_notify_datapath_link_fail(bdp_comp->bondp_ctx, wr_entry->target_vjetty,
                                                (int)send_idx, (int)target_idx);
         /* Update active link after failover is finished. */
-        bondp_health_update_active_idx(bdp_comp->bondp_ctx, wr_entry->v_conn->target_vjetty, new_send_idx);
+        bondp_health_update_active_idx(bdp_comp->bondp_ctx, wr_entry->target_vjetty, new_send_idx);
 
         bool is_primary_failover = (bdp_comp->active_count > 0 &&
                                     send_idx == (uint32_t)bdp_comp->active_indices[0] &&
@@ -840,7 +803,7 @@ static cr_convert_ret_t handle_send_cr_with_store(bondp_context_t *bdp_ctx, bond
                 .cr_status = 0,
                 .new_active_idx = -1,
                 .bdp_jetty = NULL,
-                .bdp_tjetty = wr_entry->v_conn->target_vjetty,
+                .bdp_tjetty = wr_entry->target_vjetty,
             };
             bondp_notify_health_event(bdp_comp->bondp_ctx, BONDP_HEALTH_EVENT_FALLBACK_TASK_KICK, &event_info);
         }
@@ -857,7 +820,7 @@ CONVERT_CR:
     cr->local_id = get_comp_urma_jetty_id(bdp_comp)->id;
     cr->user_ctx = wr_entry->user_ctx;
 
-    convert_jfs_pwr_to_vwr_resend(&wr_entry->wr, &wr_entry->v_conn->target_vjetty->v_tjetty);
+    convert_jfs_pwr_to_vwr_resend(&wr_entry->wr, &wr_entry->target_vjetty->v_tjetty);
     release_vwr_use_cnt(&wr_entry->wr);
     free_jfs_wr(&wr_entry->wr);
     (void)pthread_spin_lock(&bdp_comp->send_jfc->wr_lock);
diff --git a/src/urma/lib/urma/bond/bondp_datapath_convert.c b/src/urma/lib/urma/bond/bondp_datapath_convert.c
index 977ce8284a2a376b6ea95346bca389f137ebd52c..2bc52d1b6bb617ef0a9bc30b7b73327cc0460838 100644
--- a/src/urma/lib/urma/bond/bondp_datapath_convert.c
+++ b/src/urma/lib/urma/bond/bondp_datapath_convert.c
@@ -237,7 +237,7 @@ static void map_fadd_vwr_to_path(urma_jfs_wr_t *send_wr, int send_idx, int targe
 }
 
 urma_status_t convert_jfs_vwr_to_pwr(urma_jfs_wr_t *wr, int send_idx, int target_idx,
-                                     bondp_comp_t *bdp_comp, bdp_v_conn_t *v_conn)
+                                     bondp_comp_t *bdp_comp)
 {
     uint64_t opcode_tag = 0;
 
@@ -255,10 +255,10 @@ urma_status_t convert_jfs_vwr_to_pwr(urma_jfs_wr_t *wr, int send_idx, int target
             wr->opcode = URMA_OPC_SEND_IMM;
             wr->send.imm_data = encode_imm_data(
                 opcode_tag,
-                v_conn->msn,
+                bdp_comp->msn,
                 bdp_comp->v_jetty.jetty_id.id,
                 wr->send.imm_data);
-            v_conn->msn = (v_conn->msn + 1) % BONDP_MAX_BITMAP_SIZE;
+            bdp_comp->msn = (bdp_comp->msn + 1) % BONDP_MAX_BITMAP_SIZE;
 
             map_send_vwr_to_path(wr, send_idx, target_idx);
             return URMA_SUCCESS;
@@ -269,10 +269,10 @@ urma_status_t convert_jfs_vwr_to_pwr(urma_jfs_wr_t *wr, int send_idx, int target
                 opcode_tag = URMA_CR_OPC_WRITE_WITH_IMM;
                 wr->rw.notify_data = encode_imm_data(
                     opcode_tag,
-                    v_conn->msn,
+                    bdp_comp->msn,
                     bdp_comp->v_jetty.jetty_id.id,
                     wr->rw.notify_data);
-                v_conn->msn = (v_conn->msn + 1) % BONDP_MAX_BITMAP_SIZE;
+                bdp_comp->msn = (bdp_comp->msn + 1) % BONDP_MAX_BITMAP_SIZE;
             }
             map_write_vwr_to_path(wr, send_idx, target_idx);
             return URMA_SUCCESS;
diff --git a/src/urma/lib/urma/bond/bondp_datapath_convert.h b/src/urma/lib/urma/bond/bondp_datapath_convert.h
index 8eba69f88ec31276355908c8e6498f5444afee13..57f7e717fbf6bd497cc488d7b2f212e08d67666f 100644
--- a/src/urma/lib/urma/bond/bondp_datapath_convert.h
+++ b/src/urma/lib/urma/bond/bondp_datapath_convert.h
@@ -35,7 +35,7 @@ void free_jfr_wr(urma_jfr_wr_t *wr);
 void free_jfs_wr(urma_jfs_wr_t *wr);
 
 urma_status_t convert_jfs_vwr_to_pwr(urma_jfs_wr_t *wr, int send_idx, int target_idx,
-                                     bondp_comp_t *bdp_comp, bdp_v_conn_t *v_conn);
+                                     bondp_comp_t *bdp_comp);
 
 void convert_jfs_pwr_to_vwr_resend(urma_jfs_wr_t *wr, urma_target_jetty_t *vtjetty);
 
diff --git a/src/urma/lib/urma/bond/bondp_types.h b/src/urma/lib/urma/bond/bondp_types.h
index b91a0791adfc9afe8b65dcfa2168b2430059f4e4..d9ae4ac72dd9a44c98f16377f31f0c330924ed38 100644
--- a/src/urma/lib/urma/bond/bondp_types.h
+++ b/src/urma/lib/urma/bond/bondp_types.h
@@ -235,6 +235,7 @@ typedef struct bondp_comp {
     bondp_jfc_t *send_jfc;
     pthread_spinlock_t send_lock;
     bool valid[URMA_UBAGG_DEV_MAX_NUM];
+    uint32_t msn;
     urma_target_seg_t *check_tseg[URMA_UBAGG_DEV_MAX_NUM];
 #ifndef __cplusplus
     atomic_uint sqe_cnt[URMA_UBAGG_DEV_MAX_NUM];
diff --git a/src/urma/lib/urma/bond/bondp_wr_buf.c b/src/urma/lib/urma/bond/bondp_wr_buf.c
index f1104741ff41e5405098739bee6a0ae0f6ac4e66..97697f92b31f47b7396c6fa6a7fbc77532b87146 100644
--- a/src/urma/lib/urma/bond/bondp_wr_buf.c
+++ b/src/urma/lib/urma/bond/bondp_wr_buf.c
@@ -51,7 +51,7 @@ void wr_buf_uninit(wr_buf_t *buf)
 
         if (entry_hdr->entry_type == WR_BUF_ENTRY_JFS) {
             jfs_wr_entry_t *entry = (jfs_wr_entry_t *)__wr_buf_idx(buf, idx);
-            convert_jfs_pwr_to_vwr_resend(&entry->wr, &entry->v_conn->target_vjetty->v_tjetty);
+            convert_jfs_pwr_to_vwr_resend(&entry->wr, &entry->target_vjetty->v_tjetty);
             release_vwr_use_cnt(&entry->wr);
             free_jfs_wr(&entry->wr);
         } else if (entry_hdr->entry_type == WR_BUF_ENTRY_JFR) {
diff --git a/src/urma/lib/urma/bond/bondp_wr_buf.h b/src/urma/lib/urma/bond/bondp_wr_buf.h
index 0a08b3dc4a401736ada123b9ed1d7cb91db54729..2d33f0ead471490319ac1d0e15fc041f069d1d67 100644
--- a/src/urma/lib/urma/bond/bondp_wr_buf.h
+++ b/src/urma/lib/urma/bond/bondp_wr_buf.h
@@ -14,7 +14,7 @@
 #include <stdint.h>
 
 struct bondp_comp;
-struct bondp_v_connection;
+struct bondp_target_jetty;
 
 typedef enum wr_buf_entry_type {
     WR_BUF_ENTRY_NONE = 0,
@@ -33,7 +33,7 @@ typedef struct jfs_wr_entry {
     urma_jfs_wr_t wr;
     uint64_t user_ctx;
     struct bondp_comp *bdp_comp;
-    struct bondp_v_connection *v_conn;
+    struct bondp_target_jetty *target_vjetty;
     uint32_t send_idx;
     uint32_t target_idx;
 } jfs_wr_entry_t;
